// jison/lib/jison.js
// Append elements in b if not already found in a (assumes b and are individually unique)
function union(a, b) {
    var s = Object.create(null);
    for (var i = 0; i < a.length; i++) {
        s[a[i]] = 1;
    }
    for (var i = 0; i < b.length; i++) {
        if (!s[b[i]]) {
            a.push(b[i]);
        }
    }
    return a;
}

var version = require('../package.json').version;

var Jison = exports.Jison = exports;

Jison.print = console.log;

Jison.Parser = (function() {

    function Nonterminal(symbol) {
        this.symbol = symbol;
        this.productions = [];
        this.first = [];
        this.follows = [];
        this.nullable = false;
    }

    function Production(symbol, handle, id) {
        this.symbol = symbol;
        this.handle = handle;
        this.nullable = false;
        this.id = id;
        this.first = [];
        this.precedence = 0;
    }

    var generator = {};

    generator.constructor = function Jison_Generator(grammar, opt) {
        var options = Object.assign({}, grammar.options, opt);
        this.terms = {};
        this.operators = {};
        this.productions = [];
        this.conflicts = 0;
        this.resolutions = [];
        this.options = options;
        this.parseParams = grammar.parseParams;
        this.yy = {}; // accessed as yy free variable in the parser/lexer actions

        // source included in semantic action execution scope
        if (grammar.actionInclude) {
            if (typeof grammar.actionInclude === 'function') {
                grammar.actionInclude = String(grammar.actionInclude)
                    .replace(/^\s*function \(\) \{/, '')
                    .replace(/\}\s*$/, '');
            }
            this.actionInclude = grammar.actionInclude;
        }
        this.moduleInclude = grammar.moduleInclude || '';

        // Debug functionality removed

        this.processGrammar(grammar);
    };

    generator.processGrammar = function processGrammarDef(grammar) {
        var bnf = grammar.bnf,
            tokens = grammar.tokens,
            nonterminals = this.nonterminals = {},
            productions = this.productions,
            self = this;

        if (tokens) {
            if (typeof tokens === 'string') {
                tokens = tokens.trim().split(' ');
            } else {
                tokens = tokens.slice(0);
            }
        }

        var symbols = this.symbols = [];
        var operators = this.operators = processOperators(grammar.operators);
        this.buildProductions(bnf, productions, nonterminals, symbols, operators);

        if (tokens && this.terminals.length !== tokens.length) {
            self.trace("Warning: declared tokens differ from tokens found in rules.");
            self.trace(this.terminals);
            self.trace(tokens);
        }

        // augment the grammar
        this.augmentGrammar(grammar);
    };

    generator.augmentGrammar = function augmentGrammar(grammar) {
        if (this.productions.length === 0) {
            throw new Error("Grammar error: must have at least one rule.");
        }
        // use specified start symbol, or default to first user defined production
        this.startSymbol = grammar.start || grammar.startSymbol || this.productions[0].symbol;
        if (!this.nonterminals[this.startSymbol]) {
            throw new Error("Grammar error: startSymbol must be a non-terminal found in your grammar.");
        }
        this.EOF = "$end";

        // augment the grammar
        var acceptProduction = new Production('$accept', [this.startSymbol, '$end'], 0);
        this.productions.unshift(acceptProduction);
        this.symbols.unshift("$accept", this.EOF);
        this.symbols_.$accept = 0;
        this.symbols_[this.EOF] = 1;
        this.terminals.unshift(this.EOF);

        this.nonterminals.$accept = new Nonterminal("$accept");
        this.nonterminals.$accept.productions.push(acceptProduction);
        this.nonterminals[this.startSymbol].follows.push(this.EOF);
    };
    function processOperators(ops) {
        if (!ops) return {};
        var operators = {};
        for (var i = 0, k, prec; prec = ops[i]; i++) {
            for (k = 1; k < prec.length; k++) {
                operators[prec[k]] = {
                    precedence: i + 1,
                    assoc: prec[0]
                };
            }
        }
        return operators;
    }

    generator.buildProductions = function buildProductions(bnf, productions, nonterminals, symbols, operators) {
        var actions = [
            '/* this == yyval */',
            this.actionInclude || '',
            'var $0 = $$.length - 1;',
            'switch (yystate) {'
        ];
        var actionGroups = {};
        var prods, symbol;
        var productions_ = [0];
        var symbolId = 1;
        var symbols_ = {};

        var her = false; // has error recovery

        function addSymbol(s) {
            if (s && !symbols_[s]) {
                symbols_[s] = ++symbolId;
                symbols.push(s);
            }
        }

        // add error symbol; will be third symbol, or "2" ($accept, $end, error)
        addSymbol("error");

        for (symbol in bnf) {
            if (!bnf.hasOwnProperty(symbol)) continue;

            addSymbol(symbol);
            nonterminals[symbol] = new Nonterminal(symbol);

            if (typeof bnf[symbol] === 'string') {
                prods = bnf[symbol].split(/\s*\|\s*/g);
            } else {
                prods = bnf[symbol].slice(0);
            }

            prods.forEach(buildProduction);
        }

        for (var action in actionGroups) {
            actions.push(actionGroups[action].join(' '), action, 'break;');
        }

        var sym, terms = [], terms_ = {};
        for (var id in symbols_) {
            if (symbols_.hasOwnProperty(id)) {
                sym = symbols_[id];
                if (!nonterminals[sym]) {
                    terms.push(sym);
                    terms_[id] = sym;
                }
            }
        }

        this.hasErrorRecovery = her;

        this.terminals = terms;
        this.terminals_ = terms_;
        this.symbols_ = symbols_;

        this.productions_ = productions_;
        actions.push('}');

        actions = actions.join("\n")
            .replace(/YYABORT/g, 'return false')
            .replace(/YYACCEPT/g, 'return true');

        var parameters = "yytext, yyleng, yylineno, yy, yystate /* action[1] */, $$ /* vstack */, _$ /* lstack */";
        if (this.parseParams) {
            parameters += ', ' + this.parseParams.join(', ');
        }

        this.performAction = "function anonymous(" + parameters + ") {\n" + actions + "\n}";

        function buildProduction(handle) {
            var r, rhs, i;
            if (handle.constructor === Array) {
                rhs = (typeof handle[0] === 'string') ?
                    handle[0].trim().split(' ') :
                    handle[0].slice(0);

                for (i = 0; i < rhs.length; i++) {
                    if (rhs[i] === 'error') her = true;
                    if (!symbols_[rhs[i]]) {
                        addSymbol(rhs[i]);
                    }
                }

                if (typeof handle[1] === 'string' || handle.length == 3) {

                    var label = 'case ' + (productions.length + 1) + ':',
                        action = handle[1];

                    // replace named semantic values ($nonterminal)
                    if (action.match(/[$@][a-zA-Z][a-zA-Z0-9_]*/)) {
                        var count = {},
                            names = {};
                        for (i = 0; i < rhs.length; i++) {
                            // check for aliased names, e.g., id[alias]
                            var rhs_i = rhs[i].match(/\[[a-zA-Z][a-zA-Z0-9_-]*\]/);
                            if (rhs_i) {
                                rhs_i = rhs_i[0].substr(1, rhs_i[0].length - 2);
                                rhs[i] = rhs[i].substr(0, rhs[i].indexOf('['));
                            } else {
                                rhs_i = rhs[i];
                            }

                            if (names[rhs_i]) {
                                names[rhs_i + (++count[rhs_i])] = i + 1;
                            } else {
                                names[rhs_i] = i + 1;
                                names[rhs_i + "1"] = i + 1;
                                count[rhs_i] = 1;
                            }
                        }
                        action = action.replace(/\$([a-zA-Z][a-zA-Z0-9_]*)/g, function(str, pl) {
                            return names[pl] ? '$' + names[pl] : str;
                        }).replace(/@([a-zA-Z][a-zA-Z0-9_]*)/g, function(str, pl) {
                            return names[pl] ? '@' + names[pl] : str;
                        });
                    }

                    action = action
                        // replace references to $$ with this.$, and @$ with this._$
                        .replace(/([^'"])\$\$|^\$\$/g, '$1this.$').replace(/@[0$]/g, "this._$")
                        // replace semantic value references ($n) with stack value (stack[n])
                        .replace(/\$(-?\d+)/g, function(_, n) {
                            return "$$[$0" + (parseInt(n, 10) - rhs.length || '') + "]";
                        })
                        // same as above for location references (@n)
                        .replace(/@(-?\d+)/g, function(_, n) {
                            return "_$[$0" + (n - rhs.length || '') + "]";
                        });

                    if (action in actionGroups) {
                        actionGroups[action].push(label);
                    } else {
                        actionGroups[action] = [label];
                    }

                    // done with aliases; strip them.
                    rhs = rhs.map(function(e, i) {
                        return e.replace(/\[[a-zA-Z_][a-zA-Z0-9_-]*\]/g, '');
                    });
                    r = new Production(symbol, rhs, productions.length + 1);

                    if (handle[2] && operators[handle[2].prec]) {
                        r.precedence = operators[handle[2].prec].precedence;
                    }
                } else {
                    // no action -> don't care about aliases; strip them.
                    rhs = rhs.map(function(e, i) {
                        return e.replace(/\[[a-zA-Z_][a-zA-Z0-9_-]*\]/g, '');
                    });

                    r = new Production(symbol, rhs, productions.length + 1);
                    if (operators[handle[1].prec]) {
                        r.precedence = operators[handle[1].prec].precedence;
                    }
                }
            } else {
                // no action -> don't care about aliases; strip them.
                handle = handle.replace(/\[[a-zA-Z_][a-zA-Z0-9_-]*\]/g, '');
                rhs = handle.trim().split(' ');
                for (i = 0; i < rhs.length; i++) {
                    if (rhs[i] === 'error') her = true;
                    if (!symbols_[rhs[i]]) {
                        addSymbol(rhs[i]);
                    }
                }
                r = new Production(symbol, rhs, productions.length + 1);
            }

            if (r.precedence === 0) {

                for (i = r.handle.length - 1; i >= 0; i--) {
                    if (!(r.handle[i] in nonterminals) && r.handle[i] in operators) {
                        r.precedence = operators[r.handle[i]].precedence;
                    }
                }
            }

            productions.push(r);
            productions_.push([symbols_[r.symbol], r.handle[0] === '' ? 0 : r.handle.length]);
            nonterminals[symbol].productions.push(r);
        }
    };

    // noop. implemented in debug mixin
    generator.trace = function trace() {};

    generator.warn = function warn() {
        var args = Array.prototype.slice.call(arguments, 0);
        Jison.print.call(null, args.join(""));
    };

    generator.error = function error(msg) {
        throw new Error(msg);
    };

    // Generator debug mixin removed
    var lookaheadMixin = {};

    lookaheadMixin.computeLookaheads = function computeLookaheads() {
        // Lookahead debug mixin removed
        this.computeLookaheads = function() {};
        this.nullableSets();
        this.firstSets();
        this.followSets();
    };

    // calculate follow sets based on first and nullable
    lookaheadMixin.followSets = function followSets() {
        var productions = this.productions,
            nonterminals = this.nonterminals,
            self = this,
            cont = true;

        // loop until no further changes have been made
        while (cont) {
            cont = false;

            productions.forEach(function Follow_prod_forEach(production, k) {
                // q is used in Simple LALR algorithm determine follows in context
                var q;
                var ctx = !!self.go_;

                var set = [], oldcount;
                for (var i = 0, t; t = production.handle[i]; ++i) {
                    if (!nonterminals[t]) continue;

                    // for Simple LALR algorithm, self.go_ checks if
                    if (ctx) {
                        q = self.go_(production.symbol, production.handle.slice(0, i));
                    }
                    var bool = !ctx || q === parseInt(self.nterms_[t], 10);

                    if (i === production.handle.length + 1 && bool) {
                        set = nonterminals[production.symbol].follows;
                    } else {
                        var part = production.handle.slice(i + 1);

                        set = self.first(part);
                        if (self.nullable(part) && bool) {
                            set.push.apply(set, nonterminals[production.symbol].follows);
                        }
                    }
                    oldcount = nonterminals[t].follows.length;
                    union(nonterminals[t].follows, set);
                    if (oldcount !== nonterminals[t].follows.length) {
                        cont = true;
                    }
                }
            });
        }
    };

    // return the FIRST set of a symbol or series of symbols
    lookaheadMixin.first = function first(symbol) {
        if (symbol === '') {
            return [];
        } else if (symbol instanceof Array) {
            var firsts = [];
            for (var i = 0, t; t = symbol[i]; ++i) {
                if (!this.nonterminals[t]) {
                    if (firsts.indexOf(t) === -1) {
                        firsts.push(t);
                    }
                } else {
                    union(firsts, this.nonterminals[t].first);
                }
                if (!this.nullable(t)) {
                    break;
                }
            }
            return firsts;
        // terminal
        } else if (!this.nonterminals[symbol]) {
            return [symbol];
        // nonterminal
        } else {
            return this.nonterminals[symbol].first;
        }
    };

    // fixed-point calculation of FIRST sets
    lookaheadMixin.firstSets = function firstSets() {
        var productions = this.productions,
            nonterminals = this.nonterminals,
            self = this,
            cont = true,
            symbol, firsts;

        // loop until no further changes have been made
        while (cont) {
            cont = false;

            productions.forEach(function FirstSets_forEach(production, k) {
                var firsts = self.first(production.handle);
                if (firsts.length !== production.first.length) {
                    production.first = firsts;
                    cont = true;
                }
            });

            for (symbol in nonterminals) {
                firsts = [];
                nonterminals[symbol].productions.forEach(function(production) {
                    union(firsts, production.first);
                });
                if (firsts.length !== nonterminals[symbol].first.length) {
                    nonterminals[symbol].first = firsts;
                    cont = true;
                }
            }
        }
    };

    // fixed-point calculation of NULLABLE
    lookaheadMixin.nullableSets = function nullableSets() {
        var firsts = this.firsts = {},
            nonterminals = this.nonterminals,
            self = this,
            cont = true;

        // loop until no further changes have been made
        while (cont) {
            cont = false;

            // check if each production is nullable
            this.productions.forEach(function(production, k) {
                if (!production.nullable) {
                    for (var i = 0, n = 0, t; t = production.handle[i]; ++i) {
                        if (self.nullable(t)) n++;
                    }
                    if (n === i) { // production is nullable if all tokens are nullable
                        production.nullable = cont = true;
                    }
                }
            });

            //check if each symbol is nullable
            for (var symbol in nonterminals) {
                if (!this.nullable(symbol)) {
                    for (var i = 0, production; production = nonterminals[symbol].productions[i]; i++) {
                        if (production.nullable) {
                            nonterminals[symbol].nullable = cont = true;
                        }
                    }
                }
            }
        }
    };

    // check if a token or series of tokens is nullable
    lookaheadMixin.nullable = function nullable(symbol) {
        if (symbol === '') {
            return true;
        } else if (symbol instanceof Array) {
            for (var i = 0, t; t = symbol[i]; ++i) {
                if (!this.nullable(t)) {
                    return false;
                }
            }
            return true;
        // terminal
        } else if (!this.nonterminals[symbol]) {
            return false;
        // nonterminal
        } else {
            return this.nonterminals[symbol].nullable;
        }
    };

    // Lookahead debug mixin removed
    var lrGeneratorMixin = {};

    lrGeneratorMixin.buildTable = function buildTable() {
        // LR generator debug mixin removed
        this.states = this.canonicalCollection();
        this.table = this.parseTable(this.states);
        this.defaultActions = findDefaults(this.table);
    };

    function Item(production, dot, f, predecessor) {
        this.production = production;
        this.dotPosition = dot || 0;
        this.follows = f || [];
        this.predecessor = predecessor;
        this.id = parseInt(production.id + 'a' + this.dotPosition, 36);
        this.markedSymbol = this.production.handle[this.dotPosition];
    }

    Item.prototype.eq = function(e) {
        return e.id === this.id;
    };

    lrGeneratorMixin.Item = Item;

    lrGeneratorMixin.ItemSet = function ItemSet() {
        // Array-like behavior
        this._items = [];
        this.length = 0;

        // Initialize with arguments if provided
        if (arguments.length) {
            this._items = [].slice.call(arguments, 0);
            this.length = this._items.length;
        }

        // ItemSet-specific properties
        this.reductions = [];
        this.goes = {};
        this.edges = {};
        this.shifts = false;
        this.inadequate = false;
        this.hash_ = {};

        // Build hash for existing items
        for (var i = this._items.length - 1; i >= 0; i--) {
            this.hash_[this._items[i].id] = true;
        }
    };

    lrGeneratorMixin.ItemSet.prototype.concat = function concat(set) {
        var a = set._items || set;
        for (var i = a.length - 1; i >= 0; i--) {
            this.hash_[a[i].id] = true;
        }
        this._items.push.apply(this._items, a);
        this.length = this._items.length;
        return this;
    };

    lrGeneratorMixin.ItemSet.prototype.push = function(item) {
        this.hash_[item.id] = true;
        this._items.push(item);
        this.length = this._items.length;
        return this.length;
    };

    lrGeneratorMixin.ItemSet.prototype.contains = function(item) {
        return this.hash_[item.id];
    };

    lrGeneratorMixin.ItemSet.prototype.forEach = function(callback) {
        return this._items.forEach(callback);
    };

    lrGeneratorMixin.ItemSet.prototype.valueOf = function toValue() {
        var v = this._items.map(function(a) {
            return a.id;
        }).sort().join('|');
        this.valueOf = function toValue_inner() {
            return v;
        };
        return v;
    };

    lrGeneratorMixin.closureOperation = function closureOperation(itemSet /*, closureSet*/) {
        var closureSet = new this.ItemSet();
        var self = this;

        var set = itemSet,
            itemQueue, syms = {};

        do {
            itemQueue = [];
            closureSet.concat(set);
            set.forEach(function CO_set_forEach(item) {
                var symbol = item.markedSymbol;

                // if token is a non-terminal, recursively add closures
                if (symbol && self.nonterminals[symbol]) {
                    if (!syms[symbol]) {
                        self.nonterminals[symbol].productions.forEach(function CO_nt_forEach(production) {
                            var newItem = new self.Item(production, 0);
                            if (!closureSet.contains(newItem)) {
                                itemQueue.push(newItem);
                            }
                        });
                        syms[symbol] = true;
                    }
                } else if (!symbol) {
                    // reduction
                    closureSet.reductions.push(item);
                    closureSet.inadequate = closureSet.reductions.length > 1 || closureSet.shifts;
                } else {
                    // shift
                    closureSet.shifts = true;
                    closureSet.inadequate = closureSet.reductions.length > 0;
                }
            });

            set = itemQueue;

        } while (itemQueue.length > 0);

        return closureSet;
    };

    lrGeneratorMixin.gotoOperation = function gotoOperation(itemSet, symbol) {
        var gotoSet = new this.ItemSet(),
            self = this;

        itemSet.forEach(function goto_forEach(item, n) {
            if (item.markedSymbol === symbol) {
                gotoSet.push(new self.Item(item.production, item.dotPosition + 1, item.follows, n));
            }
        });

        return gotoSet.length === 0 ? gotoSet : this.closureOperation(gotoSet);
    };

    /* Create unique set of item sets */
    lrGeneratorMixin.canonicalCollection = function canonicalCollection() {
        var item1 = new this.Item(this.productions[0], 0, [this.EOF]);
        var firstState = this.closureOperation(new this.ItemSet(item1)),
            states = [firstState],
            marked = 0,
            self = this,
            itemSet;

        states.has = {};
        states.has[firstState] = 0;

        while (marked !== states.length) {
            itemSet = states[marked];
            marked++;
            itemSet.forEach(function CC_itemSet_forEach(item) {
                if (item.markedSymbol && item.markedSymbol !== self.EOF) {
                    self.canonicalCollectionInsert(item.markedSymbol, itemSet, states, marked - 1);
                }
            });
        }

        return states;
    };

    // Pushes a unique state into the que. Some parsing algorithms may perform additional operations
    lrGeneratorMixin.canonicalCollectionInsert = function canonicalCollectionInsert(symbol, itemSet, states, stateNum) {
        var g = this.gotoOperation(itemSet, symbol);
        if (!g.predecessors) {
            g.predecessors = {};
        }
        // add g to que if not empty or duplicate
        if (g.length > 0) {
            var gv = g.valueOf(),
                i = states.has[gv];
            if (i === -1 || typeof i === 'undefined') {
                states.has[gv] = states.length;
                itemSet.edges[symbol] = states.length; // store goto transition for table
                states.push(g);
                g.predecessors[symbol] = [stateNum];
            } else {
                itemSet.edges[symbol] = i; // store goto transition for table
                states[i].predecessors[symbol].push(stateNum);
            }
        }
    };

    var NONASSOC = 0;

    lrGeneratorMixin.parseTable = function parseTable(itemSets) {
        var states = [],
            nonterminals = this.nonterminals,
            operators = this.operators,
            conflictedStates = {}, // array of [state, token] tuples
            self = this,
            s = 1, // shift
            r = 2, // reduce
            a = 3; // accept

        // for each item set
        itemSets.forEach(function(itemSet, k) {
            var state = states[k] = {};
            var action, stackSymbol;

            // set shift and goto actions
            for (stackSymbol in itemSet.edges) {
                itemSet.forEach(function(item, j) {
                    // find shift and goto actions
                    if (item.markedSymbol == stackSymbol) {
                        var gotoState = itemSet.edges[stackSymbol];
                        if (nonterminals[stackSymbol]) {
                            // store state to go to after a reduce
                            state[self.symbols_[stackSymbol]] = gotoState;
                        } else {
                            state[self.symbols_[stackSymbol]] = [s, gotoState];
                        }
                    }
                });
            }

            // set accept action
            itemSet.forEach(function(item, j) {
                if (item.markedSymbol == self.EOF) {
                    // accept
                    state[self.symbols_[self.EOF]] = [a];
                }
            });

            var allterms = self.lookAheads ? false : self.terminals;

            // set reductions and resolve potential conflicts
            itemSet.reductions.forEach(function(item, j) {
                // if parser uses lookahead, only enumerate those terminals
                var terminals = allterms || self.lookAheads(itemSet, item);

                terminals.forEach(function(stackSymbol) {
                    action = state[self.symbols_[stackSymbol]];
                    var op = operators[stackSymbol];

                    // Reading a terminal and current position is at the end of a production, try to reduce
                    if (action || action && action.length) {
                        var sol = resolveConflict(item.production, op, [r, item.production.id], action[0] instanceof Array ? action[0] : action);
                        self.resolutions.push([k, stackSymbol, sol]);
                        if (sol.bydefault) {
                            self.conflicts++;
                            self.warn('Conflict in grammar: multiple actions possible when lookahead token is ', stackSymbol, ' in state ', k, "\n- ", printAction(sol.r, self), "\n- ", printAction(sol.s, self));
                            conflictedStates[k] = true;
                            if (self.options.noDefaultResolve) {
                                if (!(action[0] instanceof Array)) {
                                    action = [action];
                                }
                                action.push(sol.r);
                            }
                        } else {
                            action = sol.action;
                        }
                    } else {
                        action = [r, item.production.id];
                    }
                    if (action && action.length) {
                        state[self.symbols_[stackSymbol]] = action;
                    } else if (action === NONASSOC) {
                        state[self.symbols_[stackSymbol]] = undefined;
                    }
                });
            });
        });

        if (self.conflicts > 0) {
            self.warn("\nStates with conflicts:");
            for (var state in conflictedStates) {
                if (conflictedStates.hasOwnProperty(state)) {
                    self.warn('State ' + state);
                    self.warn('  ', itemSets[state].join("\n  "));
                }
            }
        }

        return states;
    };

    // find states with only one action, a reduction
    function findDefaults(states) {
        var defaults = {};
        states.forEach(function(state, k) {
            var i = 0;
            for (var act in state) {
                if ({}.hasOwnProperty.call(state, act)) i++;
            }

            if (i === 1 && state[act][0] === 2) {
                // only one action in state and it's a reduction
                defaults[k] = state[act];
            }
        });

        return defaults;
    }

    // resolves shift-reduce and reduce-reduce conflicts
    function resolveConflict(production, op, reduce, shift) {
        var sln = {
                production: production,
                operator: op,
                r: reduce,
                s: shift
            },
            s = 1, // shift
            r = 2, // reduce
            a = 3; // accept

        if (shift[0] === r) {
            sln.msg = "Resolve R/R conflict (use first production declared in grammar.)";
            sln.action = shift[1] < reduce[1] ? shift : reduce;
            if (shift[1] !== reduce[1]) sln.bydefault = true;
            return sln;
        }

        if (production.precedence === 0 || !op) {
            sln.msg = "Resolve S/R conflict (shift by default.)";
            sln.bydefault = true;
            sln.action = shift;
        } else if (production.precedence < op.precedence) {
            sln.msg = "Resolve S/R conflict (shift for higher precedent operator.)";
            sln.action = shift;
        } else if (production.precedence === op.precedence) {
            if (op.assoc === "right") {
                sln.msg = "Resolve S/R conflict (shift for right associative operator.)";
                sln.action = shift;
            } else if (op.assoc === "left") {
                sln.msg = "Resolve S/R conflict (reduce for left associative operator.)";
                sln.action = reduce;
            } else if (op.assoc === "nonassoc") {
                sln.msg = "Resolve S/R conflict (no action for non-associative operator.)";
                sln.action = NONASSOC;
            }
        } else {
            sln.msg = "Resolve conflict (reduce for higher precedent production.)";
            sln.action = reduce;
        }

        return sln;
    }

    lrGeneratorMixin.generate = function parser_generate(opt) {
        opt = Object.assign({}, this.options, opt);
        var code = "";

        // check for illegal identifier
        if (!opt.moduleName || !opt.moduleName.match(/^[A-Za-z_$][A-Za-z0-9_$]*$/)) {
            opt.moduleName = "parser";
        }
        switch (opt.moduleType) {
            case "js":
                code = this.generateModule(opt);
                break;
            default:
                code = this.generateCommonJSModule(opt);
                break;
        }

        return code;
    };

    lrGeneratorMixin.generateCommonJSModule = function generateCommonJSModule(opt) {
        opt = Object.assign({}, this.options, opt);
        var moduleName = opt.moduleName || "parser";
        var out = this.generateModule(opt) +
            "\n\n\nif (typeof require !== 'undefined' && typeof exports !== 'undefined') {" +
            "\nexports.parser = " + moduleName + ";" +
            "\nexports.Parser = " + moduleName + ".Parser;" +
            "\nexports.parse = function () { return " + moduleName + ".parse.apply(" + moduleName + ", arguments); };" +
            "\nexports.main = " + String(opt.moduleMain || commonjsMain) + ";" +
            "\nif (typeof module !== 'undefined' && require.main === module) {\n" +
            "  exports.main(process.argv.slice(1));\n}" +
            "\n}";

        return out;
    };

    lrGeneratorMixin.generateModule = function generateModule(opt) {
        opt = Object.assign({}, this.options, opt);
        var moduleName = opt.moduleName || "parser";
        var now = new Date().toISOString();
        var out = "/* parser generated by jison " + version + " on " + now + " */\n";
        out += (moduleName.match(/\./) ? moduleName : "var " + moduleName) +
            " = " + this.generateModuleExpr();

        return out;
    };

    lrGeneratorMixin.generateModuleExpr = function generateModuleExpr() {
        var out = '';
        var module = this.generateModule_();

        out += "(function(){\n";
        out += module.commonCode;
        out += "\nvar parser = " + module.moduleCode;
        out += "\n" + this.moduleInclude;
        out += "\nfunction Parser () {\n  this.yy = {};\n}\n" +
            "Parser.prototype = parser;" +
            "parser.Parser = Parser;" +
            "\nreturn new Parser;\n})();";

        return out;
    };

    // Token stack support removed - not used by CoffeeScript
    lrGeneratorMixin.generateModule_ = function generateModule_() {
        var parseFn = String(parser.parse);

        // Token stack support removed

        nextVariableId = 0;
        var tableCode = this.generateTableCode(this.table);
        var commonCode = tableCode.commonCode;
        var moduleCode = "{";
        moduleCode += [
            "trace: " + String(this.trace || parser.trace),
            "yy: {}",
            "symbols_: " + JSON.stringify(this.symbols_),
            "terminals_: " + JSON.stringify(this.terminals_).replace(/"([0-9]+)":/g, "$1:"),
            "productions_: " + JSON.stringify(this.productions_),
            "performAction: " + String(this.performAction),
            "table: " + tableCode.moduleCode,
            "defaultActions: " + JSON.stringify(this.defaultActions).replace(/"([0-9]+)":/g, "$1:"),
            "parseError: " + String(this.parseError || (this.hasErrorRecovery ? traceParseError : parser.parseError)),
            "parse: " + parseFn
        ].join(",\n");
        moduleCode += "};";

        return {
            commonCode: commonCode,
            moduleCode: moduleCode
        };
    };

    // Generate code that represents the specified parser table
    lrGeneratorMixin.generateTableCode = function(table) {
        var moduleCode = JSON.stringify(table);
        var variables = [createObjectCode];
        moduleCode = moduleCode.replace(/"([0-9]+)"(?=:)/g, "$1");

        // Replace objects with identical values by function calls
        moduleCode = moduleCode.replace(/\{\d+:[^\}]+,\d+:[^\}]+\}/g, function(object) {
            var value, frequentValue, key, keys = {},
                keyCount, maxKeyCount = 0,
                keyValue, keyValues = [],
                keyValueMatcher = /(\d+):([^:]+)(?=,\d+:|\})/g;

            while ((keyValue = keyValueMatcher.exec(object))) {
                key = keyValue[1];
                value = keyValue[2];
                keyCount = 1;

                if (!(value in keys)) {
                    keys[value] = [key];
                } else {
                    keyCount = keys[value].push(key);
                }
                if (keyCount > maxKeyCount) {
                    maxKeyCount = keyCount;
                    frequentValue = value;
                }
            }
            if (maxKeyCount > 1) {
                for (value in keys) {
                    if (value !== frequentValue) {
                        for (var k = keys[value], i = 0, l = k.length; i < l; i++) {
                            keyValues.push(k[i] + ':' + value);
                        }
                    }
                }
                keyValues = keyValues.length ? ',{' + keyValues.join(',') + '}' : '';
                object = 'o([' + keys[frequentValue].join(',') + '],' + frequentValue + keyValues + ')';
            }
            return object;
        });

        var list;
        var lists = {};
        var listMatcher = /\[[0-9,]+\]/g;

        while (list = listMatcher.exec(moduleCode)) {
            lists[list] = (lists[list] || 0) + 1;
        }

        moduleCode = moduleCode.replace(listMatcher, function(list) {
            var listId = lists[list];
            if (typeof listId === 'number') {
                if (listId === 1) {
                    lists[list] = listId = list;
                } else {
                    lists[list] = listId = createVariable();
                    variables.push(listId + '=' + list);
                }
            }
            return listId;
        });

        return {
            commonCode: 'var ' + variables.join(',') + ';',
            moduleCode: moduleCode
        };
    };
    var createObjectCode = 'o=function(k,v,o,l){' +
        'for(o=o||{},l=k.length;l--;o[k[l]]=v);' +
        'return o}';
    function createVariable() {
        var id = nextVariableId++;
        var name = '$V';

        do {
            name += variableTokens[id % variableTokensLength];
            id = ~~(id / variableTokensLength);
        } while (id !== 0);

        return name;
    }

    var nextVariableId = 0;
    var variableTokens = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ_$';
    var variableTokensLength = variableTokens.length;

    // default main method for generated commonjs modules
    function commonjsMain(args) {
        if (!args[1]) {
            console.log('Usage: ' + args[0] + ' FILE');
            process.exit(1);
        }
        var source = require('fs').readFileSync(require('path').normalize(args[1]), "utf8");
        return exports.parser.parse(source);
    }

    // LR generator debug mixin removed

    var parser = {};

    lrGeneratorMixin.createParser = function createParser() {
        var p = eval(this.generateModuleExpr());

        // for debugging
        p.productions = this.productions;

        var self = this;

        function bind(method) {
            return function() {
                self.lexer = p.lexer;
                return self[method].apply(self, arguments);
            };
        }

        // backwards compatability
        p.lexer = this.lexer;
        p.generate = bind('generate');
        p.generateModule = bind('generateModule');
        p.generateCommonJSModule = bind('generateCommonJSModule');

        return p;
    };

    parser.trace = generator.trace;
    parser.warn = generator.warn;
    parser.error = generator.error;

    function traceParseError(err, hash) {
        this.trace(err);
    }

    function parseError(str, hash) {
        if (hash.recoverable) {
            this.trace(str);
        } else {
            var error = new Error(str);
            error.hash = hash;
            throw error;
        }
    }

    parser.parseError = lrGeneratorMixin.parseError = parseError;

    parser.parse = function parse(input) {
        var self = this,
            stack = [0],
            tstack = [], // token stack
            vstack = [null], // semantic value stack
            lstack = [], // location stack
            table = this.table,
            yytext = '',
            yylineno = 0,
            yyleng = 0,
            recovering = 0,
            TERROR = 2,
            EOF = 1;

        var args = lstack.slice.call(arguments, 1);

        //this.reductionCount = this.shiftCount = 0;

        var lexer = Object.create(this.lexer);
        var sharedState = {
            yy: {}
        };

        // copy state
        for (var k in this.yy) {
            if (Object.prototype.hasOwnProperty.call(this.yy, k)) {
                sharedState.yy[k] = this.yy[k];
            }
        }

        lexer.setInput(input, sharedState.yy);
        sharedState.yy.lexer = lexer;
        sharedState.yy.parser = this;
        if (typeof lexer.yylloc == 'undefined') {
            lexer.yylloc = {};
        }
        var yyloc = lexer.yylloc;
        lstack.push(yyloc);

        var ranges = lexer.options && lexer.options.ranges;

        if (typeof sharedState.yy.parseError === 'function') {
            this.parseError = sharedState.yy.parseError;
        } else {
            this.parseError = Object.getPrototypeOf(this).parseError;
        }

        function popStack(n) {
            stack.length = stack.length - 2 * n;
            vstack.length = vstack.length - n;
            lstack.length = lstack.length - n;
        }

        _token_stack:
        var lex = function() {
            var token;
            token = lexer.lex() || EOF;
            // if token isn't its numeric value, convert
            if (typeof token !== 'number') {
                token = self.symbols_[token] || token;
            }
            return token;
        }

        var symbol, preErrorSymbol, state, action, a, r, yyval = {},
            p, len, newState, expected;

        while (true) {
            // retreive state number from top of stack
            state = stack[stack.length - 1];

            // use default actions if available
            if (this.defaultActions[state]) {
                action = this.defaultActions[state];
            } else {
                if (symbol === null || typeof symbol == 'undefined') {
                    symbol = lex();
                }
                // read action for current state and first input
                action = table[state] && table[state][symbol];
            }

            _handle_error:
            // handle parse error
            if (typeof action === 'undefined' || !action.length || !action[0]) {
                var errStr = '';

                if (!recovering) {
                    // Report error
                    expected = [];
                    for (p in table[state]) {
                        if (this.terminals_[p] && p > TERROR) {
                            expected.push("'" + this.terminals_[p] + "'");
                        }
                    }
                    if (lexer.showPosition) {
                        errStr = 'Parse error on line ' + (yylineno + 1) + ":\n" + lexer.showPosition() + "\nExpecting " + expected.join(', ') + ", got '" + (this.terminals_[symbol] || symbol) + "'";
                    } else {
                        errStr = 'Parse error on line ' + (yylineno + 1) + ": Unexpected " +
                            (symbol == EOF ? "end of input" :
                                ("'" + (this.terminals_[symbol] || symbol) + "'"));
                    }
                    this.parseError(errStr, {
                        text: lexer.match,
                        token: this.terminals_[symbol] || symbol,
                        line: lexer.yylineno,
                        loc: yyloc,
                        expected: expected
                    });
                }

                throw new Error(errStr);
            }

            // this shouldn't happen, unless resolve defaults are off
            if (action[0] instanceof Array && action.length > 1) {
                throw new Error('Parse Error: multiple actions possible at state: ' + state + ', token: ' + symbol);
            }

            switch (action[0]) {
                case 1: // shift
                    //this.shiftCount++;

                    stack.push(symbol);
                    vstack.push(lexer.yytext);
                    lstack.push(lexer.yylloc);
                    stack.push(action[1]); // push state
                    symbol = null;
                    if (!preErrorSymbol) { // normal execution/no error
                        yyleng = lexer.yyleng;
                        yytext = lexer.yytext;
                        yylineno = lexer.yylineno;
                        yyloc = lexer.yylloc;
                        if (recovering > 0) {
                            recovering--;
                        }
                    } else {
                        // error just occurred, resume old lookahead f/ before error
                        symbol = preErrorSymbol;
                        preErrorSymbol = null;
                    }
                    break;

                case 2:
                    // reduce
                    //this.reductionCount++;

                    len = this.productions_[action[1]][1];

                    // perform semantic action
                    yyval.$ = vstack[vstack.length - len]; // default to $$ = $1
                    // default location, uses first token for firsts, last for lasts
                    yyval._$ = {
                        first_line: lstack[lstack.length - (len || 1)].first_line,
                        last_line: lstack[lstack.length - 1].last_line,
                        first_column: lstack[lstack.length - (len || 1)].first_column,
                        last_column: lstack[lstack.length - 1].last_column
                    };
                    if (ranges) {
                        yyval._$.range = [lstack[lstack.length - (len || 1)].range[0], lstack[lstack.length - 1].range[1]];
                    }
                    r = this.performAction.apply(yyval, [yytext, yyleng, yylineno, sharedState.yy, action[1], vstack, lstack].concat(args));

                    if (typeof r !== 'undefined') {
                        return r;
                    }

                    // pop off stack
                    if (len) {
                        stack = stack.slice(0, -1 * len * 2);
                        vstack = vstack.slice(0, -1 * len);
                        lstack = lstack.slice(0, -1 * len);
                    }

                    stack.push(this.productions_[action[1]][0]); // push nonterminal (reduce)
                    vstack.push(yyval.$);
                    lstack.push(yyval._$);
                    // goto new state = table[STATE][NONTERMINAL]
                    newState = table[stack[stack.length - 2]][stack[stack.length - 1]];
                    stack.push(newState);
                    break;

                case 3:
                    // accept
                    return true;
            }
        }

        return true;
    };

    parser.init = function parser_init(dict) {
        this.table = dict.table;
        this.defaultActions = dict.defaultActions;
        this.performAction = dict.performAction;
        this.productions_ = dict.productions_;
        this.symbols_ = dict.symbols_;
        this.terminals_ = dict.terminals_;
    };
    function LALRGenerator(grammar, options) {
        // Call base constructor
        generator.constructor.call(this, grammar, options);

        this.type = "LALR(1)";

        // Debug mixins removed

        options = options || {};
        this.states = this.canonicalCollection();
        this.terms_ = {};

        var newg = this.newg = Object.create(lookaheadMixin);
        Object.assign(newg, {
            oldg: this,
            trace: this.trace,
            nterms_: {},
            go_: function(r, B) {
                r = r.split(":")[0]; // grab state #
                B = B.map(function(b) {
                    return b.slice(b.indexOf(":") + 1);
                });
                return this.oldg.go(r, B);
            }
        });
        newg.nonterminals = {};
        newg.productions = [];

        this.inadequateStates = [];

        // if true, only lookaheads in inadequate states are computed (faster, larger table)
        // if false, lookaheads for all reductions will be computed (slower, smaller table)
        this.onDemandLookahead = options.onDemandLookahead || false;

        this.buildNewGrammar();
        newg.computeLookaheads();
        this.unionLookaheads();

        this.table = this.parseTable(this.states);
        this.defaultActions = findDefaults(this.table);
    }

    // Set up prototype chain: LALRGenerator inherits from generator, lookaheadMixin, and lrGeneratorMixin
    LALRGenerator.prototype = Object.create(generator);
    Object.assign(LALRGenerator.prototype, lookaheadMixin, lrGeneratorMixin);
    LALRGenerator.prototype.constructor = LALRGenerator;

    LALRGenerator.prototype.lookAheads = function LALR_lookaheads(state, item) {
        return (!!this.onDemandLookahead && !state.inadequate) ? this.terminals : item.follows;
    };

    LALRGenerator.prototype.go = function LALR_go(p, w) {
        var q = parseInt(p, 10);
        for (var i = 0; i < w.length; i++) {
            q = this.states[q].edges[w[i]] || q;
        }
        return q;
    };

    LALRGenerator.prototype.goPath = function LALR_goPath(p, w) {
        var q = parseInt(p, 10),
            t,
            path = [];
        for (var i = 0; i < w.length; i++) {
            t = w[i] ? q + ":" + w[i] : '';
            if (t) this.newg.nterms_[t] = q;
            path.push(t);
            q = this.states[q].edges[w[i]] || q;
            this.terms_[t] = w[i];
        }
        return {
            path: path,
            endState: q
        };
    };

    // every disjoint reduction of a nonterminal becomes a produciton in G'
    LALRGenerator.prototype.buildNewGrammar = function LALR_buildNewGrammar() {
        var self = this,
            newg = this.newg;

        this.states.forEach(function(state, i) {
            state.forEach(function(item) {
                if (item.dotPosition === 0) {
                    // new symbols are a combination of state and transition symbol
                    var symbol = i + ":" + item.production.symbol;
                    self.terms_[symbol] = item.production.symbol;
                    newg.nterms_[symbol] = i;
                    if (!newg.nonterminals[symbol]) {
                        newg.nonterminals[symbol] = new Nonterminal(symbol);
                    }
                    var pathInfo = self.goPath(i, item.production.handle);
                    var p = new Production(symbol, pathInfo.path, newg.productions.length);
                    newg.productions.push(p);
                    newg.nonterminals[symbol].productions.push(p);

                    // store the transition that get's 'backed up to' after reduction on path
                    var handle = item.production.handle.join(' ');
                    var goes = self.states[pathInfo.endState].goes;
                    if (!goes[handle]) {
                        goes[handle] = [];
                    }
                    goes[handle].push(symbol);
                }
            });
            if (state.inadequate) {
                self.inadequateStates.push(i);
            }
        });
    };

    LALRGenerator.prototype.unionLookaheads = function LALR_unionLookaheads() {
        var self = this,
            newg = this.newg,
            states = !!this.onDemandLookahead ? this.inadequateStates : this.states;

        states.forEach(function union_states_forEach(i) {
            var state = typeof i === 'number' ? self.states[i] : i,
                follows = [];
            if (state.reductions.length) {
                state.reductions.forEach(function union_reduction_forEach(item) {
                    var follows = {};
                    for (var k = 0; k < item.follows.length; k++) {
                        follows[item.follows[k]] = true;
                    }
                    state.goes[item.production.handle.join(' ')].forEach(function reduction_goes_forEach(symbol) {
                        newg.nonterminals[symbol].follows.forEach(function goes_follows_forEach(symbol) {
                            var terminal = self.terms_[symbol];
                            if (!follows[terminal]) {
                                follows[terminal] = true;
                                item.follows.push(terminal);
                            }
                        });
                    });
                });
            }
        });
    };

    exports.LALRGenerator = LALRGenerator;

    // LALR generator debug mixin removed

    Jison.Generator = function Jison_Generator(g, options) {
        var opt = Object.assign({}, g.options, options);
        return new LALRGenerator(g, opt);
    };

    return function Parser(g, options) {
        var gen = Jison.Generator(g, options);
        return gen.createParser();
    };

})();
