// jison/lib/jison.js
var Jison = exports.Jison = exports;

Jison.version = require('../package.json').version;

class Nonterminal {
  constructor(symbol) {
    this.symbol = symbol;
    this.productions = [];
    this.first = [];
    this.follows = [];
    this.nullable = false;
  }

  toString() {
    var str = this.symbol + "\n";
    str += (this.nullable ? 'nullable' : 'not nullable');
    str += "\nFirsts: " + this.first.join(', ');
    str += "\nFollows: " + this.first.join(', ');
    str += "\nProductions:\n  " + this.productions.join('\n  ');

    return str;
  }
}

class Production {
  constructor(symbol, handle, id) {
    this.symbol = symbol;
    this.handle = handle;
    this.nullable = false;
    this.id = id;
    this.first = [];
    this.precedence = 0;
  }

  toString() {
    return this.symbol + " -> " + this.handle.join(' ');
  }
}

class Item {
  constructor(production, dot, f, predecessor) {
    this.production = production;
    this.dotPosition = dot || 0;
    this.follows = f || [];
    this.predecessor = predecessor;
    this.id = parseInt(production.id + 'a' + this.dotPosition, 36);
    this.markedSymbol = this.production.handle[this.dotPosition];
  }

  remainingHandle() {
    return this.production.handle.slice(this.dotPosition + 1);
  }

  eq(e) {
    return e.id === this.id;
  }

  handleToString() {
    var handle = this.production.handle.slice(0);
    handle[this.dotPosition] = '.' + (handle[this.dotPosition] || '');
    return handle.join(' ');
  }

  toString() {
    var temp = this.production.handle.slice(0);
    temp[this.dotPosition] = '.' + (temp[this.dotPosition] || '');
    return (
      this.production.symbol +
      ' -> ' +
      temp.join(' ') +
      (this.follows.length === 0 ? '' : ' #lookaheads= ' + this.follows.join(' '))
    );
  }
}

class ItemSet {
  constructor() {
    this.map = new Map(); // id -> item
    this.reductions = [];
    this.goes = {};
    this.edges = {};
    this.shifts = false;
    this.inadequate = false;
  }

  push(item) {
    if (!this.map.has(item.id)) {
      this.map.set(item.id, item);
    }
    return this.map.size;
  }

  concat(set) {
    // Handle both ItemSet and array inputs
    const items = set.items ? set.items() : set;
    for (const item of items) {
      this.push(item);
    }
    return this;
  }

  contains(item) {
    return this.map.has(item.id);
  }

  valueOf() {
    // Memoized string representation for state deduplication
    const v = Array.from(this.map.keys()).sort().join('|');
    this.valueOf = function () {
      return v;
    };
    return v;
  }

  forEach(fn) {
    for (const item of this.map.values()) {
      fn(item);
    }
  }

  isEmpty() {
    return this.map.size === 0;
  }

  items() {
    return Array.from(this.map.values());
  }

  toString() {
    return this.items().toString();
  }
}

class Generator {

  // ==[ Constructor ]========================================================

  constructor(grammar, opt) {
    var options = Object.assign({}, grammar.options, opt);
    this.terms = {};
    this.operators = {};
    this.productions = [];
    this.conflicts = 0;
    this.resolutions = [];
    this.options = options;
    this.parseParams = grammar.parseParams;
    this.yy = {}; // accessed as yy free variable in the parser/lexer actions

    // source included in semantic action execution scope
    if (grammar.actionInclude) {
      if (typeof grammar.actionInclude === 'function') {
        grammar.actionInclude = String(grammar.actionInclude)
          .replace(/^function \(\) \{/, '')
          .replace(/\}\s*$/, '');
      }
      this.actionInclude = grammar.actionInclude;
    }
    this.moduleInclude = grammar.moduleInclude || '';

    this.processGrammar(grammar);

    if (grammar.lex) {
      var Lexer = require('jison-lex');
      this.lexer = new Lexer(grammar.lex, null, this.terminals_);
    }
  }

  // ==[ Setup ]==============================================================

  processGrammar(grammar) {
    var bnf = grammar.bnf,
      tokens = grammar.tokens,
      nonterminals = this.nonterminals = {},
      productions = this.productions;

    if (tokens) {
      if (typeof tokens === 'string') {
        tokens = tokens.trim().split(' ');
      } else {
        tokens = tokens.slice(0);
      }
    }

    var symbols = this.symbols = [];

    // calculate precedence of operators
    var operators = this.operators = processOperators(grammar.operators);

    // build productions from cfg
    this.buildProductions(bnf, productions, nonterminals, symbols, operators);

    // augment the grammar
    this.augmentGrammar(grammar);
  }

  augmentGrammar(grammar) {
    if (this.productions.length === 0) {
      throw new Error("Grammar error: must have at least one rule.");
    }
    // use specified start symbol, or default to first user defined production
    this.startSymbol = grammar.start || grammar.startSymbol || this.productions[0].symbol;
    if (!this.nonterminals[this.startSymbol]) {
      throw new Error("Grammar error: startSymbol must be a non-terminal found in your grammar.");
    }
    this.EOF = "$end";

    // augment the grammar
    var acceptProduction = new Production('$accept', [this.startSymbol, '$end'], 0);
    this.productions.unshift(acceptProduction);

    // prepend parser tokens
    this.symbols.unshift("$accept", this.EOF);
    this.symbols_[this.EOF] = 1;
    this.terminals.unshift(this.EOF);

    this.nonterminals.$accept = new Nonterminal("$accept");
    this.nonterminals.$accept.productions.push(acceptProduction);

    // add follow $ to start symbol
    this.nonterminals[this.startSymbol].follows.push(this.EOF);
  }

  // ==[ Grammar building ]===================================================

  buildProductions(bnf, productions, nonterminals, symbols, operators) {
    var actions = [
      '/* this == yyval */',
      this.actionInclude || '',
      'var $0 = $$.length - 1;',
      'switch (yystate) {'
    ];
    var actionGroups = {};
    var prods, symbol;
    var productions_ = [0];
    var symbolId = 1;
    var symbols_ = {};

    var her = false; // has error recovery

    function addSymbol(s) {
      if (s && !symbols_[s]) {
        symbols_[s] = ++symbolId;
        symbols.push(s);
      }
    }

    // add error symbol; will be third symbol, or "2" ($accept, $end, error)
    addSymbol("error");

    for (symbol in bnf) {
      if (!bnf.hasOwnProperty(symbol)) continue;

      addSymbol(symbol);
      nonterminals[symbol] = new Nonterminal(symbol);

      if (typeof bnf[symbol] === 'string') {
        prods = bnf[symbol].split(/\s*\|\s*/g);
      } else {
        prods = bnf[symbol].slice(0);
      }

      prods.forEach(buildProduction);
    }
    for (var action in actionGroups)
      actions.push(actionGroups[action].join(' '), action, 'break;');

    var sym, terms = [], terms_ = {};
    each(symbols_, function (id, sym) {
      if (!nonterminals[sym]) {
        terms.push(sym);
        terms_[id] = sym;
      }
    });

    this.hasErrorRecovery = her;

    this.terminals = terms;
    this.terminals_ = terms_;
    this.symbols_ = symbols_;

    this.productions_ = productions_;
    actions.push('}');

    actions = actions.join("\n")
      .replace(/YYABORT/g, 'return false')
      .replace(/YYACCEPT/g, 'return true');

    var parameters = "yytext, yyleng, yylineno, yy, yystate /* action[1] */, $$ /* vstack */, _$ /* lstack */";
    if (this.parseParams) parameters += ', ' + this.parseParams.join(', ');

    this.performAction = "function anonymous(" + parameters + ") {\n" + actions + "\n}";

    function buildProduction(handle) {
      var r, rhs, i;
      if (handle.constructor === Array) {
        rhs = (typeof handle[0] === 'string') ?
          handle[0].trim().split(' ') :
          handle[0].slice(0);

        for (i = 0; i < rhs.length; i++) {
          if (rhs[i] === 'error') her = true;
          if (!symbols_[rhs[i]]) {
            addSymbol(rhs[i]);
          }
        }

        if (typeof handle[1] === 'string' || handle.length == 3) {
          // semantic action specified
          var label = 'case ' + (productions.length + 1) + ':', action = handle[1];

          // replace named semantic values ($nonterminal)
          if (action.match(/[$@][a-zA-Z][a-zA-Z0-9_]*/)) {
            var count = {},
              names = {};
            for (i = 0; i < rhs.length; i++) {
              // check for aliased names, e.g., id[alias]
              var rhs_i = rhs[i].match(/\[[a-zA-Z][a-zA-Z0-9_-]*\]/);
              if (rhs_i) {
                rhs_i = rhs_i[0].substr(1, rhs_i[0].length - 2);
                rhs[i] = rhs[i].substr(0, rhs[i].indexOf('['));
              } else {
                rhs_i = rhs[i];
              }

              if (names[rhs_i]) {
                names[rhs_i + (++count[rhs_i])] = i + 1;
              } else {
                names[rhs_i] = i + 1;
                names[rhs_i + "1"] = i + 1;
                count[rhs_i] = 1;
              }
            }
            action = action.replace(/\$([a-zA-Z][a-zA-Z0-9_]*)/g, function (str, pl) {
              return names[pl] ? '$' + names[pl] : str;
            }).replace(/@([a-zA-Z][a-zA-Z0-9_]*)/g, function (str, pl) {
              return names[pl] ? '@' + names[pl] : str;
            });
          }
          action = action
            // replace references to $$ with this.$, and @$ with this._$
            .replace(/([^'"])\$\$|^\$\$/g, '$1this.$').replace(/@[0$]/g, "this._$")

            // replace semantic value references ($n) with stack value (stack[n])
            .replace(/\$(-?\d+)/g, function (_, n) {
              return "$$[$0" + (parseInt(n, 10) - rhs.length || '') + "]";
            })
            // same as above for location references (@n)
            .replace(/@(-?\d+)/g, function (_, n) {
              return "_$[$0" + (n - rhs.length || '') + "]";
            });
          if (action in actionGroups) actionGroups[action].push(label);
          else actionGroups[action] = [label];

          // done with aliases; strip them.
          rhs = rhs.map(function (e, i) { return e.replace(/\[[a-zA-Z_][a-zA-Z0-9_-]*\]/g, '') });
          r = new Production(symbol, rhs, productions.length + 1);
          // precedence specified also
          if (handle[2] && operators[handle[2].prec]) {
            r.precedence = operators[handle[2].prec].precedence;
          }
        } else {
          // no action -> don't care about aliases; strip them.
          rhs = rhs.map(function (e, i) { return e.replace(/\[[a-zA-Z_][a-zA-Z0-9_-]*\]/g, '') });
          // only precedence specified
          r = new Production(symbol, rhs, productions.length + 1);
          if (operators[handle[1].prec]) {
            r.precedence = operators[handle[1].prec].precedence;
          }
        }
      } else {
        // no action -> don't care about aliases; strip them.
        handle = handle.replace(/\[[a-zA-Z_][a-zA-Z0-9_-]*\]/g, '');
        rhs = handle.trim().split(' ');
        for (i = 0; i < rhs.length; i++) {
          if (rhs[i] === 'error') her = true;
          if (!symbols_[rhs[i]]) {
            addSymbol(rhs[i]);
          }
        }
        r = new Production(symbol, rhs, productions.length + 1);
      }
      if (r.precedence === 0) {
        // set precedence
        for (i = r.handle.length - 1; i >= 0; i--) {
          if (!(r.handle[i] in nonterminals) && r.handle[i] in operators) {
            r.precedence = operators[r.handle[i]].precedence;
          }
        }
      }

      productions.push(r);
      productions_.push([symbols_[r.symbol], r.handle[0] === '' ? 0 : r.handle.length]);
      nonterminals[symbol].productions.push(r);
    }
  }

  buildTable() {
    this.states = this.canonicalCollection();
    this.table = this.parseTable(this.states);
    this.defaultActions = findDefaults(this.table);
  }

  canonicalCollection() {
    var item1 = new Item(this.productions[0], 0, [this.EOF]);
    var firstSet = new ItemSet();
    firstSet.push(item1);
    var firstState = this.closureOperation(firstSet),
      states = [firstState],
      marked = 0,
      self = this,
      itemSet;

    states.has = {};
    states.has[firstState.valueOf()] = 0;

    while (marked !== states.length) {
      itemSet = states[marked]; marked++;
      itemSet.forEach((item) => {
        if (item.markedSymbol && item.markedSymbol !== self.EOF)
          self.canonicalCollectionInsert(item.markedSymbol, itemSet, states, marked - 1);
      });
    }

    return states;
  }

  canonicalCollectionInsert(symbol, itemSet, states, stateNum) {
    var g = this.gotoOperation(itemSet, symbol);
    if (!g.predecessors)
      g.predecessors = {};
    // add g to queue if not empty or duplicate
    if (!g.isEmpty()) {
      var gv = g.valueOf(),
        i = states.has[gv];
      if (i === -1 || typeof i === 'undefined') {
        states.has[gv] = states.length;
        itemSet.edges[symbol] = states.length; // store goto transition for table
        states.push(g);
        g.predecessors[symbol] = [stateNum];
      } else {
        itemSet.edges[symbol] = i; // store goto transition for table
        states[i].predecessors[symbol].push(stateNum);
      }
    }
  }

  gotoOperation(itemSet, symbol) {
    var gotoSet = new ItemSet(),
      self = this;

    itemSet.forEach((item) => {
      if (item.markedSymbol === symbol) {
        gotoSet.push(new Item(item.production, item.dotPosition + 1, item.follows, item.predecessor));
      }
    });

    return gotoSet.isEmpty() ? gotoSet : this.closureOperation(gotoSet);
  }

  closureOperation(itemSet /*, closureSet*/) {
    const closureSet = new ItemSet();
    let set = itemSet,
      itemQueue, syms = {};

    do {
      itemQueue = new ItemSet();
      closureSet.concat(set);
      set.forEach((item) => {
        const symbol = item.markedSymbol;

        // if token is a non-terminal, recursively add closures
        if (symbol && this.nonterminals[symbol]) {
          if (!syms[symbol]) {
            this.nonterminals[symbol].productions.forEach((production) => {
              const newItem = new Item(production, 0);
              if (!closureSet.contains(newItem))
                itemQueue.push(newItem);
            });
            syms[symbol] = true;
          }
        } else if (!symbol) {
          // reduction
          closureSet.reductions.push(item);
          closureSet.inadequate = closureSet.reductions.length > 1 || closureSet.shifts;
        } else {
          // shift
          closureSet.shifts = true;
          closureSet.inadequate = closureSet.reductions.length > 0;
        }
      });

      set = itemQueue;
    } while (!itemQueue.isEmpty());

    return closureSet;
  }

  parseTable(itemSets) {
    const states = [];
    const nonterminals = this.nonterminals;
    const operators = this.operators;
    const conflictedStates = {}; // array of [state, token] tuples
    const s = 1, r = 2, a = 3; // shift, reduce, accept

    // for each item set
    itemSets.forEach((itemSet, k) => {
      const state = states[k] = {};
      let action, stackSymbol;

      // set shift and goto actions
      for (stackSymbol in itemSet.edges) {
        itemSet.forEach((item, j) => {
          // find shift and goto actions
          if (item.markedSymbol == stackSymbol) {
            const gotoState = itemSet.edges[stackSymbol];
            if (nonterminals[stackSymbol]) {
              // store state to go to after a reduce
              //this.trace(k, stackSymbol, 'g'+gotoState);
              state[this.symbols_[stackSymbol]] = gotoState;
            } else {
              //this.trace(k, stackSymbol, 's'+gotoState);
              state[this.symbols_[stackSymbol]] = [s, gotoState];
            }
          }
        });
      }

      // set accept action
      itemSet.forEach((item, j) => {
        if (item.markedSymbol == this.EOF) {
          // accept
          state[this.symbols_[this.EOF]] = [a];
          //this.trace(k, this.EOF, state[this.EOF]);
        }
      });

      const allterms = this.lookAheads ? false : this.terminals;

      // set reductions and resolve potential conflicts
      itemSet.reductions.forEach((item, j) => {
        // if parser uses lookahead, only enumerate those terminals
        const terminals = allterms || this.lookAheads(itemSet, item);

        terminals.forEach((stackSymbol) => {
          action = state[this.symbols_[stackSymbol]];
          const op = operators[stackSymbol];

          // Reading a terminal and current position is at the end of a production, try to reduce
          if (action || (action && action.length)) {
            const sol = resolveConflict(item.production, op, [r, item.production.id], action[0] instanceof Array ? action[0] : action);
            this.resolutions.push([k, stackSymbol, sol]);
            if (sol.bydefault) {
              this.conflicts++;
              throw new Error('Grammar ambiguous when lookahead is ' + stackSymbol + ' in state ' + k);
            } else {
              action = sol.action;
            }
          } else {
            action = [r, item.production.id];
          }
          if (action && action.length) {
            state[this.symbols_[stackSymbol]] = action;
          } else if (action === NONASSOC) {
            state[this.symbols_[stackSymbol]] = undefined;
          }
        });
      });
    });

    if (this.conflicts > 0) {
      let conflictDetails = "\nStates with conflicts:";
      each(conflictedStates, (val, state) => {
        conflictDetails += '\nState ' + state;
        conflictDetails += '\n  ' + itemSets[state].join("\n  ");
      });
      throw new Error("Grammar conflicts: " + this.conflicts + " " + conflictDetails);
    }

    return states;
  }

  // ==[ Lookahead / Parsing algorithms ]=====================================

  // Lookahead computation methods
  computeLookaheads() {
    this.nullableSets();
    this.firstSets();
    this.followSets();
  }

  nullableSets() {
    this.firsts = {};
    const nonterminals = this.nonterminals;
    let cont = true;

    // loop until no further changes have been made
    while (cont) {
      cont = false;

      // check if each production is nullable
      this.productions.forEach((production, k) => {
        if (!production.nullable) {
          let n = 0, i = 0, t;
          for (i = 0; (t = production.handle[i]); ++i) {
            if (this.nullable(t)) n++;
          }
          if (n === i) { // production is nullable if all tokens are nullable
            production.nullable = cont = true;
          }
        }
      });

      // check if each symbol is nullable
      for (const symbol in nonterminals) {
        if (!this.nullable(symbol)) {
          for (let i = 0, production; (production = nonterminals[symbol].productions[i]); i++) {
            if (production.nullable)
              nonterminals[symbol].nullable = cont = true;
          }
        }
      }
    }
  }

  nullable(symbol) {
    // epsilon
    if (symbol === '') {
      return true;
      // RHS
    } else if (Array.isArray(symbol)) {
      for (let i = 0, t; (t = symbol[i]); ++i) {
        if (!this.nullable(t))
          return false;
      }
      return true;
      // terminal
    } else if (!this.nonterminals[symbol]) {
      return false;
      // nonterminal
    } else {
      return this.nonterminals[symbol].nullable;
    }
  }

  firstSets() {
    const productions = this.productions;
    const nonterminals = this.nonterminals;
    let cont = true;

    // loop until no further changes have been made
    while (cont) {
      cont = false;
      productions.forEach((production, k) => {
        let firsts = this.first(production.handle);
        let oldcount = nonterminals[production.symbol].first.length;
        unionArrays(nonterminals[production.symbol].first, firsts);
        if (oldcount !== nonterminals[production.symbol].first.length) {
          cont = true;
        }
      });
    }
  }

  first(symbol) {
    // epsilon
    if (symbol === '') {
      return [];
      // RHS
    } else if (symbol instanceof Array) {
      var firsts = [];
      for (var i = 0, t; t = symbol[i]; ++i) {
        if (!this.nonterminals[t]) {
          if (firsts.indexOf(t) === -1)
            firsts.push(t);
        } else {
          unionArrays(firsts, this.nonterminals[t].first);
        }
        if (!this.nullable(t))
          break;
      }
      return firsts;
      // terminal
    } else if (!this.nonterminals[symbol]) {
      return [symbol];
      // nonterminal
    } else {
      return this.nonterminals[symbol].first;
    }
  }

  followSets() {
    const productions = this.productions;
    const nonterminals = this.nonterminals;
    let cont = true;

    // loop until no further changes have been made
    while (cont) {
      cont = false;

      productions.forEach((production, k) => {
        // q is used in Simple LALR algorithm to determine follows in context
        let q;
        let ctx = !!this.go_;

        let set = [], oldcount;
        for (let i = 0, t; t = production.handle[i]; ++i) {
          if (!nonterminals[t]) continue;

          // For Simple LALR algorithm, this.go_ checks if
          if (ctx) {
            q = this.go_(production.symbol, production.handle.slice(0, i));
            var bool = !ctx || q === parseInt(this.nterms_[t], 10);

            if (i === production.handle.length + 1 && bool) {
              set = nonterminals[production.symbol].follows;
            } else {
              var part = production.handle.slice(i + 1);
              set = this.first(part);
              if (this.nullable(part) && bool) {
                set.push.apply(set, nonterminals[production.symbol].follows);
              }
            }
          } else {
            var part = production.handle.slice(i + 1);
            set = this.first(part);
            if (this.nullable(part)) {
              set.push.apply(set, nonterminals[production.symbol].follows);
            }
          }
          oldcount = nonterminals[t].follows.length;
          unionArrays(nonterminals[t].follows, set);
          if (oldcount !== nonterminals[t].follows.length) {
            cont = true;
          }
        }
      });
    }
  }

  // ==[ Output/Generation ]==================================================

  generate(opt) {
    opt = Object.assign({}, this.options, opt);

    // check for illegal identifier
    if (!opt.moduleName || !opt.moduleName.match(/^[A-Za-z_$][A-Za-z0-9_$]*$/)) {
      opt.moduleName = "parser";
    }
    switch (opt.moduleType) {
      case "js":
        return this.generateModule(opt);
      default:
        return this.generateCommonJSModule(opt);
    }
  }

  generateCommonJSModule(opt) {
    opt = Object.assign({}, this.options, opt);
    var moduleName = opt.moduleName || "parser";

    return this.generateModule(opt)
      + "\n\n\nif (typeof require !== 'undefined' && typeof exports !== 'undefined') {"
      + "\nexports.parser = " + moduleName + ";"
      + "\nexports.Parser = " + moduleName + ".Parser;"
      + "\nexports.parse = function () { return " + moduleName + ".parse.apply(" + moduleName + ", arguments); };"
      + "\nexports.main = function() {};"
      + "\n}";
  }

  generateModule(opt) {
    opt = Object.assign({}, this.options, opt);
    var moduleName = opt.moduleName || "parser";

    var out = "/* parser generated by rip-parser " + Jison.version + " */\n";
    out += (moduleName.match(/\./) ? moduleName : "var " + moduleName) + " = ";
    out += this.generateModuleExpr();

    return out;
  }

  generateModuleExpr() {
    var out = "(function(){\n";
    out += "\nvar parser = {";
    out += [
      "trace: function trace () { }",
      "yy: {}",
      "symbols_: " + JSON.stringify(this.symbols_),
      "terminals_: " + JSON.stringify(this.terminals_).replace(/\"(\d+)\":/g, '$1:'),
      "productions_: " + JSON.stringify(this.productions_),
      "performAction: " + String(this.performAction),
      "table: " + JSON.stringify(this.table).replaceAll('"', ''),
      "defaultActions: " + JSON.stringify(this.defaultActions).replaceAll('"', ''),
      "parseError: " + String(this.parseError || parser.parseError),
      "parse: " + String(parser.parse)
    ].join(",\n");
    out += "};\n";
    out += this.moduleInclude;
    if (this.lexer && this.lexer.generateModule) {
      out += this.lexer.generateModule();
      out += "\nparser.lexer = lexer;";
    }
    out += "\nfunction Parser () {\n  this.yy = {};\n}\n";
    out += "Parser.prototype = parser;";
    out += "parser.Parser = Parser;";
    out += "\nreturn new Parser;\n})();";

    return out;
  }

  createParser() {
    var p = eval(this.generateModuleExpr());

    // for debugging
    p.productions = this.productions;

    var self = this;
    function bind(method) {
      return function () {
        self.lexer = p.lexer;
        return self[method].apply(self, arguments);
      };
    }

    // backwards compatability
    p.lexer = this.lexer;
    p.generate = bind('generate');
    p.generateCommonJSModule = bind('generateCommonJSModule');
    p.generateModule = bind('generateModule');

    return p;
  }

  // ==[ Error handling / logging ]===========================================

  trace() { }

  warn() {
    var args = Array.prototype.slice.call(arguments, 0);
    throw new Error("Warning: " + args.join(""));
  }

  error(msg) {
    throw new Error(msg);
  }
}

class LALRGenerator extends Generator {
  constructor(grammar, options) {
    super(grammar, options);

    this.type = "LALR(1)";

    options = options || {};
    this.states = this.canonicalCollection();
    this.terms_ = {};

    var newg = this.newg = Object.assign(Object.create(Generator.prototype), {
      oldg: this,
      trace: this.trace,
      nterms_: {},
      go_: function (r, B) {
        r = r.split(":")[0]; // grab state #
        B = B.map(function (b) { return b.slice(b.indexOf(":") + 1); });
        return this.oldg.go(r, B);
      }
    });
    newg.nonterminals = {};
    newg.productions = [];

    this.inadequateStates = [];

    // if true, only lookaheads in inadequate states are computed (faster, larger table)
    // if false, lookaheads for all reductions will be computed (slower, smaller table)
    this.onDemandLookahead = options.onDemandLookahead || false;

    this.buildNewGrammar();
    newg.computeLookaheads();
    this.unionLookaheads();

    this.table = this.parseTable(this.states);
    this.defaultActions = findDefaults(this.table);
  }

  lookAheads(state, item) {
    return (!!this.onDemandLookahead && !state.inadequate) ? this.terminals : item.follows;
  }

  go(p, w) {
    var q = parseInt(p, 10);
    for (var i = 0; i < w.length; i++) {
      q = this.states[q].edges[w[i]] || q;
    }
    return q;
  }

  goPath(p, w) {
    var q = parseInt(p, 10), t,
      path = [];
    for (var i = 0; i < w.length; i++) {
      t = w[i] ? q + ":" + w[i] : '';
      if (t) this.newg.nterms_[t] = q;
      path.push(t);
      q = this.states[q].edges[w[i]] || q;
      this.terms_[t] = w[i];
    }
    return { path: path, endState: q };
  }

  // every disjoint reduction of a nonterminal becomes a produciton in G'
  buildNewGrammar() {
    var self = this,
      newg = this.newg;

    this.states.forEach(function (state, i) {
      state.forEach(function (item) {
        if (item.dotPosition === 0) {
          // new symbols are a combination of state and transition symbol
          var symbol = i + ":" + item.production.symbol;
          self.terms_[symbol] = item.production.symbol;
          newg.nterms_[symbol] = i;
          if (!newg.nonterminals[symbol])
            newg.nonterminals[symbol] = new Nonterminal(symbol);
          var pathInfo = self.goPath(i, item.production.handle);
          var p = new Production(symbol, pathInfo.path, newg.productions.length);
          newg.productions.push(p);
          newg.nonterminals[symbol].productions.push(p);

          // store the transition that get's 'backed up to' after reduction on path
          var handle = item.production.handle.join(' ');
          var goes = self.states[pathInfo.endState].goes;
          if (!goes[handle])
            goes[handle] = [];
          goes[handle].push(symbol);
          //self.trace('new production:',p);
        }
      });
      if (state.inadequate)
        self.inadequateStates.push(i);
    });
  }

  unionLookaheads() {
    var self = this,
      newg = this.newg,
      states = !!this.onDemandLookahead ? this.inadequateStates : this.states;

    states.forEach(function union_states_forEach(i) {
      var state = typeof i === 'number' ? self.states[i] : i,
        follows = [];
      if (state.reductions.length)
        state.reductions.forEach(function union_reduction_forEach(item) {
          var follows = {};
          for (var k = 0; k < item.follows.length; k++) {
            follows[item.follows[k]] = true;
          }
          state.goes[item.production.handle.join(' ')].forEach(function reduction_goes_forEach(symbol) {
            newg.nonterminals[symbol].follows.forEach(function goes_follows_forEach(symbol) {
              var terminal = self.terms_[symbol];
              if (!follows[terminal]) {
                follows[terminal] = true;
                item.follows.push(terminal);
              }
            });
          });
          //self.trace('unioned item', item);
        });
    });
  }
}

// iterator utility
function each(obj, func) {
  if (obj.forEach) {
    obj.forEach(func);
  } else {
    var p;
    for (p in obj) {
      if (obj.hasOwnProperty(p)) {
        func.call(obj, obj[p], p, obj);
      }
    }
  }
}

// Helper function to merge arrays without duplicates
function unionArrays(a, b) {
  var ar = {};
  for (var k = a.length - 1; k >= 0; --k) {
    ar[a[k]] = true;
  }
  for (var i = b.length - 1; i >= 0; --i) {
    if (!ar[b[i]]) {
      a.push(b[i]);
    }
  }
  return a;
}

// set precedence and associativity of operators
function processOperators(ops) {
  if (!ops) return {};
  var operators = {};
  for (var i = 0, k, prec; prec = ops[i]; i++) {
    for (k = 1; k < prec.length; k++) {
      operators[prec[k]] = { precedence: i + 1, assoc: prec[0] };
    }
  }
  return operators;
}

function findDefaults(states) {
  var defaults = {};
  states.forEach(function (state, k) {
    var i = 0;
    for (var act in state) {
      if ({}.hasOwnProperty.call(state, act)) i++;
    }

    if (i === 1 && state[act][0] === 2) {
      // only one action in state and it's a reduction
      defaults[k] = state[act];
    }
  });

  return defaults;
}

function resolveConflict(production, op, reduce, shift) {
  var sln = { production: production, operator: op, r: reduce, s: shift },
    s = 1, // shift
    r = 2, // reduce
    a = 3; // accept

  if (shift[0] === r) {
    sln.msg = "Resolve R/R conflict (use first production declared in grammar.)";
    sln.action = shift[1] < reduce[1] ? shift : reduce;
    if (shift[1] !== reduce[1]) sln.bydefault = true;
    return sln;
  }

  if (production.precedence === 0 || !op) {
    sln.msg = "Resolve S/R conflict (shift by default.)";
    sln.bydefault = true;
    sln.action = shift;
  } else if (production.precedence < op.precedence) {
    sln.msg = "Resolve S/R conflict (shift for higher precedent operator.)";
    sln.action = shift;
  } else if (production.precedence === op.precedence) {
    if (op.assoc === "right") {
      sln.msg = "Resolve S/R conflict (shift for right associative operator.)";
      sln.action = shift;
    } else if (op.assoc === "left") {
      sln.msg = "Resolve S/R conflict (reduce for left associative operator.)";
      sln.action = reduce;
    } else if (op.assoc === "nonassoc") {
      sln.msg = "Resolve S/R conflict (no action for non-associative operator.)";
      sln.action = 0; // NONASSOC
    }
  } else {
    sln.msg = "Resolve conflict (reduce for higher precedent production.)";
    sln.action = reduce;
  }

  return sln;
}

var generator = Generator.prototype;

var parser = {};

parser.trace = generator.trace;
parser.warn = generator.warn;
parser.error = generator.error;

parser.parseError = parseError = function parseError(str, hash) {
  if (hash.recoverable) {
    this.trace(str);
  } else {
    var error = new Error(str);
    error.hash = hash;
    throw error;
  }
}

parser.parse = function parse(input) {
  var self = this,
    stack = [0],
    tstack = [], // token stack
    vstack = [null], // semantic value stack
    lstack = [], // location stack
    table = this.table,
    yytext = '',
    yylineno = 0,
    yyleng = 0,
    recovering = 0,
    TERROR = 2,
    EOF = 1;

  var args = lstack.slice.call(arguments, 1);
  var lexer = Object.create(this.lexer);
  var sharedState = { yy: {} };

  // copy state
  for (var k in this.yy) {
    if (Object.prototype.hasOwnProperty.call(this.yy, k)) {
      sharedState.yy[k] = this.yy[k];
    }
  }

  lexer.setInput(input, sharedState.yy);
  sharedState.yy.lexer = lexer;
  sharedState.yy.parser = this;
  if (typeof lexer.yylloc == 'undefined') {
    lexer.yylloc = {};
  }
  var yyloc = lexer.yylloc;
  lstack.push(yyloc);

  var ranges = lexer.options && lexer.options.ranges;

  if (typeof sharedState.yy.parseError === 'function') {
    this.parseError = sharedState.yy.parseError;
  } else {
    this.parseError = Object.getPrototypeOf(this).parseError;
  }

  function popStack(n) {
    stack.length = stack.length - 2 * n;
    vstack.length = vstack.length - n;
    lstack.length = lstack.length - n;
  }

  var lex = function () {
    var token;
    token = lexer.lex() || EOF;
    // if token isn't its numeric value, convert
    if (typeof token !== 'number') {
      token = self.symbols_[token] || token;
    }
    return token;
  }

  var symbol, preErrorSymbol, state, action, a, r, yyval = {}, p, len, newState, expected;
  while (true) {
    // retreive state number from top of stack
    state = stack[stack.length - 1];

    // use default actions if available
    if (this.defaultActions[state]) {
      action = this.defaultActions[state];
    } else {
      if (symbol === null || typeof symbol == 'undefined') {
        symbol = lex();
      }
      // read action for current state and first input
      action = table[state] && table[state][symbol];
    }

    // handle parse error
    if (typeof action === 'undefined' || !action.length || !action[0]) {
      var error_rule_depth;
      var errStr = '';

      // Return the rule stack depth where the nearest error rule can be found.
      // Return FALSE when no error recovery rule was found.
      function locateNearestErrorRecoveryRule(state) {
        var stack_probe = stack.length - 1;
        var depth = 0;

        // try to recover from error
        for (; ;) {
          // check for error recovery rule in this state
          if ((TERROR.toString()) in table[state]) {
            return depth;
          }
          if (state === 0 || stack_probe < 2) {
            return false; // No suitable error recovery rule available.
          }
          stack_probe -= 2; // popStack(1): [symbol, action]
          state = stack[stack_probe];
          ++depth;
        }
      }

      if (!recovering) {
        // first see if there's any chance at hitting an error recovery rule:
        error_rule_depth = locateNearestErrorRecoveryRule(state);

        // Report error
        expected = [];
        for (p in table[state]) {
          if (this.terminals_[p] && p > TERROR) {
            expected.push("'" + this.terminals_[p] + "'");
          }
        }
        if (lexer.showPosition) {
          errStr = 'Parse error on line ' + (yylineno + 1) + ":\n" + lexer.showPosition() + "\nExpecting " + expected.join(', ') + ", got '" + (this.terminals_[symbol] || symbol) + "'";
        } else {
          errStr = 'Parse error on line ' + (yylineno + 1) + ": Unexpected " +
            (symbol == EOF ? "end of input" :
              ("'" + (this.terminals_[symbol] || symbol) + "'"));
        }
        this.parseError(errStr, {
          text: lexer.match,
          token: this.terminals_[symbol] || symbol,
          line: lexer.yylineno,
          loc: yyloc,
          expected: expected,
          recoverable: (error_rule_depth !== false)
        });
      } else if (preErrorSymbol !== EOF) {
        error_rule_depth = locateNearestErrorRecoveryRule(state);
      }

      // just recovered from another error
      if (recovering == 3) {
        if (symbol === EOF || preErrorSymbol === EOF) {
          throw new Error(errStr || 'Parsing halted while starting to recover from another error.');
        }

        // discard current lookahead and grab another
        yyleng = lexer.yyleng;
        yytext = lexer.yytext;
        yylineno = lexer.yylineno;
        yyloc = lexer.yylloc;
        symbol = lex();
      }

      // try to recover from error
      if (error_rule_depth === false) {
        throw new Error(errStr || 'Parsing halted. No suitable error recovery rule available.');
      }
      popStack(error_rule_depth);

      preErrorSymbol = (symbol == TERROR ? null : symbol); // save the lookahead token
      symbol = TERROR;         // insert generic error symbol as new lookahead
      state = stack[stack.length - 1];
      action = table[state] && table[state][TERROR];
      recovering = 3; // allow 3 real symbols to be shifted before reporting a new error
    }

    // this shouldn't happen, unless resolve defaults are off
    if (action[0] instanceof Array && action.length > 1) {
      throw new Error('Parse Error: multiple actions possible at state: ' + state + ', token: ' + symbol);
    }

    switch (action[0]) {
      case 1: // shift
        //this.shiftCount++;

        stack.push(symbol);
        vstack.push(lexer.yytext);
        lstack.push(lexer.yylloc);
        stack.push(action[1]); // push state
        symbol = null;
        if (!preErrorSymbol) { // normal execution/no error
          yyleng = lexer.yyleng;
          yytext = lexer.yytext;
          yylineno = lexer.yylineno;
          yyloc = lexer.yylloc;
          if (recovering > 0) {
            recovering--;
          }
        } else {
          // error just occurred, resume old lookahead f/ before error
          symbol = preErrorSymbol;
          preErrorSymbol = null;
        }
        break;

      case 2:
        // reduce
        //this.reductionCount++;

        len = this.productions_[action[1]][1];

        // perform semantic action
        yyval.$ = vstack[vstack.length - len]; // default to $$ = $1
        // default location, uses first token for firsts, last for lasts
        yyval._$ = {
          first_line: lstack[lstack.length - (len || 1)].first_line,
          last_line: lstack[lstack.length - 1].last_line,
          first_column: lstack[lstack.length - (len || 1)].first_column,
          last_column: lstack[lstack.length - 1].last_column
        };
        if (ranges) {
          yyval._$.range = [lstack[lstack.length - (len || 1)].range[0], lstack[lstack.length - 1].range[1]];
        }
        r = this.performAction.apply(yyval, [yytext, yyleng, yylineno, sharedState.yy, action[1], vstack, lstack].concat(args));

        if (typeof r !== 'undefined') {
          return r;
        }

        // pop off stack
        if (len) {
          stack = stack.slice(0, -1 * len * 2);
          vstack = vstack.slice(0, -1 * len);
          lstack = lstack.slice(0, -1 * len);
        }

        stack.push(this.productions_[action[1]][0]);    // push nonterminal (reduce)
        vstack.push(yyval.$);
        lstack.push(yyval._$);
        // goto new state = table[STATE][NONTERMINAL]
        newState = table[stack[stack.length - 2]][stack[stack.length - 1]];
        stack.push(newState);
        break;

      case 3:
        // accept
        return true;
    }

  }

  return true;
};

parser.init = function parser_init(dict) {
  this.table = dict.table;
  this.defaultActions = dict.defaultActions;
  this.performAction = dict.performAction;
  this.productions_ = dict.productions_;
  this.symbols_ = dict.symbols_;
  this.terminals_ = dict.terminals_;
};

exports.main = function () { };

exports.LALRGenerator = LALRGenerator;

Jison.Generator = function Jison_Generator(g, options) {
  var opt = Object.assign({}, g.options, options);
  // Only LALR(1) parser is supported
  return new LALRGenerator(g, opt);
};

// Add Parser class for legacy compatibility
class Parser extends LALRGenerator {
  constructor(grammar, options) {
    super(grammar, options);
  }
}

// CommonJS compatibility
module.exports = { Parser, Generator, LALRGenerator, Nonterminal, Production, Item, ItemSet };

return function Parser(g, options) {
  var gen = Jison.Generator(g, options);
  return gen.createParser();
};
