// jison/lib/jison.js
// LALR(1) Lookahead Generation engine (DeRemer-Pennello Algorithm)
// Creates augmented grammar to compute lookahead sets for parser states
// Unified Language class eliminates the original newg spawning pattern
//
// Original implementation by Jison team
// https://github.com/zaach/jison/blob/master/lib/jison.js

var version = require('../package.json').version;
var Jison = exports.Jison = exports;

Jison.print = console.log;

// Utility functions
function union(a, b) {
    var s = Object.create(null);
    var i, len;
    for (i = 0, len = a.length; i < len; i++) {
        s[a[i]] = 1;
    }
    for (i = 0, len = b.length; i < len; i++) {
        if (!s[b[i]]) {
            a.push(b[i]);
        }
    }
    return a;
}

// Core classes
function Nonterminal(symbol) {
    this.symbol = symbol;
    this.productions = [];
    this.first = [];
    this.follows = [];
    this.nullable = false;
}

function Production(symbol, handle, id) {
    this.symbol = symbol;
    this.handle = handle;
    this.nullable = false;
    this.id = id;
    this.first = [];
    this.precedence = 0;
}

function Item(production, dot, f, predecessor) {
    this.production = production;
    this.dotPosition = dot || 0;
    this.follows = f || [];
    this.predecessor = predecessor;
    this.id = parseInt(production.id + 'a' + this.dotPosition, 36);
    this.markedSymbol = this.production.handle[this.dotPosition];
}

function ItemSet() {
    this._items = [];
    this.length = 0;
    this.reductions = [];
    this.goes = {};
    this.edges = {};
    this.shifts = false;
    this.inadequate = false;
    this.hash_ = {};

    if (arguments.length) {
        this._items = Array.prototype.slice.call(arguments);
        this.length = this._items.length;
        for (var i = this.length - 1; i >= 0; i--) {
            this.hash_[this._items[i].id] = true;
        }
    }
}

ItemSet.prototype.concat = function(set) {
    var a = set._items || set;
    for (var i = a.length - 1; i >= 0; i--) {
        this.hash_[a[i].id] = true;
    }
    this._items.push.apply(this._items, a);
    this.length = this._items.length;
    return this;
};

ItemSet.prototype.push = function(item) {
    this.hash_[item.id] = true;
    this._items.push(item);
    this.length = this._items.length;
    return this.length;
};

ItemSet.prototype.contains = function(item) {
    return this.hash_[item.id];
};

ItemSet.prototype.valueOf = function() {
    var v = this._items.map(function(a) { return a.id; }).sort().join('|');
    this.valueOf = function() { return v; };
    return v;
};

// Unified Language class for parser generation
function Language(grammar, options) {
    options = Object.assign({}, grammar.options, options);
    this.terms = {};
    this.operators = {};
    this.productions = [];
    this.conflicts = 0;
    this.resolutions = [];
    this.options = options;
    this.parseParams = grammar.parseParams;
    this.yy = {};

    if (grammar.actionInclude) {
        if (typeof grammar.actionInclude === 'function') {
            grammar.actionInclude = String(grammar.actionInclude)
                .replace(/^\s*function \(\) \{/, '')
                .replace(/\}\s*$/, '');
        }
        this.actionInclude = grammar.actionInclude;
    }
    this.moduleInclude = grammar.moduleInclude || '';

    this.processGrammar(grammar);
    this.states = this.canonicalCollection();
    this.terms_ = {};

    // Initialize lookahead state (replaces newg creation)
    this.lookahead = {
        nterms_: {},
        nonterminals: {},
        productions: []
    };

    this.inadequateStates = [];
    this.onDemandLookahead = options.onDemandLookahead || false;

    this.buildNewGrammar();

    // Compute lookaheads in lookahead context (replaces newg.computeLookaheads())
    var savedNonterminals = this.nonterminals;
    var savedProductions = this.productions;

    this.nonterminals = this.lookahead.nonterminals;
    this.productions = this.lookahead.productions;

    this.computeLookaheads();

    this.nonterminals = savedNonterminals;
    this.productions = savedProductions;

    this.unionLookaheads();
    this.table = this.parseTable(this.states);
    this.defaultActions = this.findDefaults(this.table);
}

// Specialized go_ method for lookahead computation (from original newg)
Language.prototype.go_ = function(r, B) {
    r = r.split(":")[0];
    B = B.map(function(b) { return b.slice(b.indexOf(":") + 1); });
    return this.go(r, B);
};

Language.prototype.processGrammar = function(grammar) {
    var bnf = grammar.bnf;
    var tokens = grammar.tokens;
    var nonterminals = this.nonterminals = {};
    var productions = this.productions;
    var symbols = this.symbols = [];
    var operators = this.operators = this.processOperators(grammar.operators);

    if (tokens) {
        tokens = typeof tokens === 'string' ? tokens.trim().split(' ') : tokens.slice(0);
    }

    this.buildProductions(bnf, productions, nonterminals, symbols, operators);

    if (tokens && this.terminals.length !== tokens.length) {
        this.trace("Warning: declared tokens differ from tokens found in rules.");
    }

    this.augmentGrammar(grammar);
};

Language.prototype.processOperators = function(ops) {
    if (!ops) return {};
    var operators = {};
    for (var i = 0, prec; prec = ops[i]; i++) {
        for (var k = 1; k < prec.length; k++) {
            operators[prec[k]] = {
                precedence: i + 1,
                assoc: prec[0]
            };
        }
    }
    return operators;
};

Language.prototype.augmentGrammar = function(grammar) {
    if (this.productions.length === 0) {
        throw new Error("Grammar error: must have at least one rule.");
    }

    this.startSymbol = grammar.start || grammar.startSymbol || this.productions[0].symbol;
    if (!this.nonterminals[this.startSymbol]) {
        throw new Error("Grammar error: startSymbol must be a non-terminal found in your grammar.");
    }

    this.EOF = "$end";
    var acceptProduction = new Production('$accept', [this.startSymbol, '$end'], 0);
    this.productions.unshift(acceptProduction);
    this.symbols.unshift("$accept", this.EOF);
    this.symbols_.$accept = 0;
    this.symbols_[this.EOF] = 1;
    this.terminals.unshift(this.EOF);

    this.nonterminals.$accept = new Nonterminal("$accept");
    this.nonterminals.$accept.productions.push(acceptProduction);
    this.nonterminals[this.startSymbol].follows.push(this.EOF);
};

Language.prototype.buildProductions = function(bnf, productions, nonterminals, symbols, operators) {
    var actions = [
        '/* this == yyval */',
        this.actionInclude || '',
        'var $0 = $$.length - 1;',
        'switch (yystate) {'
    ];
    var actionGroups = {};
    var productions_ = [0];
    var symbolId = 1;
    var symbols_ = {};
    var her = false;

    function addSymbol(s) {
        if (s && !symbols_[s]) {
            symbols_[s] = ++symbolId;
            symbols.push(s);
        }
    }

    addSymbol("error");

    for (var symbol in bnf) {
        if (!bnf.hasOwnProperty(symbol)) continue;

        addSymbol(symbol);
        nonterminals[symbol] = new Nonterminal(symbol);

        var prods = typeof bnf[symbol] === 'string' ?
            bnf[symbol].split(/\s*\|\s*/g) :
            bnf[symbol].slice(0);

        prods.forEach(function(handle) {
            var r, rhs, i;
            if (handle.constructor === Array) {
                rhs = typeof handle[0] === 'string' ?
                    handle[0].trim().split(' ') :
                    handle[0].slice(0);

                for (i = 0; i < rhs.length; i++) {
                    if (rhs[i] === 'error') her = true;
                    addSymbol(rhs[i]);
                }

                if (typeof handle[1] === 'string' || handle.length === 3) {
                    var label = 'case ' + (productions.length + 1) + ':';
                    var action = handle[1];

                    // Process named semantic values
                    if (action.match(/[$@][a-zA-Z][a-zA-Z0-9_]*/)) {
                        var count = {}, names = {};
                        for (i = 0; i < rhs.length; i++) {
                            var rhs_i = rhs[i].match(/\[[a-zA-Z][a-zA-Z0-9_-]*\]/);
                            if (rhs_i) {
                                rhs_i = rhs_i[0].substr(1, rhs_i[0].length - 2);
                                rhs[i] = rhs[i].substr(0, rhs[i].indexOf('['));
                            } else {
                                rhs_i = rhs[i];
                            }

                            if (names[rhs_i]) {
                                names[rhs_i + (++count[rhs_i])] = i + 1;
                            } else {
                                names[rhs_i] = i + 1;
                                names[rhs_i + "1"] = i + 1;
                                count[rhs_i] = 1;
                            }
                        }

                        action = action.replace(/\$([a-zA-Z][a-zA-Z0-9_]*)/g, function(str, pl) {
                            return names[pl] ? '$' + names[pl] : str;
                        }).replace(/@([a-zA-Z][a-zA-Z0-9_]*)/g, function(str, pl) {
                            return names[pl] ? '@' + names[pl] : str;
                        });
                    }

                    action = action
                        .replace(/([^'"])\$\$|^\$\$/g, '$1this.$')
                        .replace(/@[0$]/g, "this._$")
                        .replace(/\$(-?\d+)/g, function(_, n) {
                            return "$$[$0" + (parseInt(n, 10) - rhs.length || '') + "]";
                        })
                        .replace(/@(-?\d+)/g, function(_, n) {
                            return "_$[$0" + (n - rhs.length || '') + "]";
                        });

                    if (actionGroups[action]) {
                        actionGroups[action].push(label);
                    } else {
                        actionGroups[action] = [label];
                    }

                    rhs = rhs.map(function(e) {
                        return e.replace(/\[[a-zA-Z_][a-zA-Z0-9_-]*\]/g, '');
                    });

                    r = new Production(symbol, rhs, productions.length + 1);
                    if (handle[2] && operators[handle[2].prec]) {
                        r.precedence = operators[handle[2].prec].precedence;
                    }
                } else {
                    rhs = rhs.map(function(e) {
                        return e.replace(/\[[a-zA-Z_][a-zA-Z0-9_-]*\]/g, '');
                    });

                    r = new Production(symbol, rhs, productions.length + 1);
                    if (operators[handle[1].prec]) {
                        r.precedence = operators[handle[1].prec].precedence;
                    }
                }
            } else {
                handle = handle.replace(/\[[a-zA-Z_][a-zA-Z0-9_-]*\]/g, '');
                rhs = handle.trim().split(' ');
                for (i = 0; i < rhs.length; i++) {
                    if (rhs[i] === 'error') her = true;
                    addSymbol(rhs[i]);
                }
                r = new Production(symbol, rhs, productions.length + 1);
            }

            if (r.precedence === 0) {
                for (i = r.handle.length - 1; i >= 0; i--) {
                    if (!(r.handle[i] in nonterminals) && r.handle[i] in operators) {
                        r.precedence = operators[r.handle[i]].precedence;
                        break;
                    }
                }
            }

            productions.push(r);
            productions_.push([symbols_[r.symbol], r.handle[0] === '' ? 0 : r.handle.length]);
            nonterminals[symbol].productions.push(r);
        });
    }

    for (var action in actionGroups) {
        actions.push(actionGroups[action].join(' '), action, 'break;');
    }

    var terms = [], terms_ = {};
    for (var id in symbols_) {
        if (symbols_.hasOwnProperty(id)) {
            var sym = symbols_[id];
            if (!nonterminals[sym]) {
                terms.push(sym);
                terms_[id] = sym;
            }
        }
    }

    this.hasErrorRecovery = her;
    this.terminals = terms;
    this.terminals_ = terms_;
    this.symbols_ = symbols_;
    this.productions_ = productions_;

    actions.push('}');
    var actionsCode = actions.join("\n")
        .replace(/YYABORT/g, 'return false')
        .replace(/YYACCEPT/g, 'return true');

    var parameters = "yytext, yyleng, yylineno, yy, yystate, $$, _$";
    if (this.parseParams) {
        parameters += ', ' + this.parseParams.join(', ');
    }

    this.performAction = "function anonymous(" + parameters + ") {\n" + actionsCode + "\n}";
};

// Lookahead computation
Language.prototype.computeLookaheads = function() {
    this.computeLookaheads = function() {};
    this.nullableSets();
    this.firstSets();
    this.followSets();
};

Language.prototype.nullableSets = function() {
    var nonterminals = this.nonterminals;
    var self = this;
    var cont = true;

    while (cont) {
        cont = false;

        this.productions.forEach(function(production) {
            if (!production.nullable) {
                var n = 0;
                for (var i = 0, t; t = production.handle[i]; ++i) {
                    if (self.nullable(t)) n++;
                }
                if (n === i) {
                    production.nullable = cont = true;
                }
            }
        });

        for (var symbol in nonterminals) {
            if (!this.nullable(symbol)) {
                for (var i = 0, production; production = nonterminals[symbol].productions[i]; i++) {
                    if (production.nullable) {
                        nonterminals[symbol].nullable = cont = true;
                        break;
                    }
                }
            }
        }
    }
};

Language.prototype.nullable = function(symbol) {
    if (symbol === '') return true;
    if (symbol instanceof Array) {
        for (var i = 0, t; t = symbol[i]; ++i) {
            if (!this.nullable(t)) return false;
        }
        return true;
    }
    return !this.nonterminals[symbol] ? false : this.nonterminals[symbol].nullable;
};

Language.prototype.firstSets = function() {
    var productions = this.productions;
    var nonterminals = this.nonterminals;
    var self = this;
    var cont = true;

    while (cont) {
        cont = false;

        productions.forEach(function(production) {
            var firsts = self.first(production.handle);
            if (firsts.length !== production.first.length) {
                production.first = firsts;
                cont = true;
            }
        });

        for (var symbol in nonterminals) {
            var firsts = [];
            nonterminals[symbol].productions.forEach(function(production) {
                union(firsts, production.first);
            });
            if (firsts.length !== nonterminals[symbol].first.length) {
                nonterminals[symbol].first = firsts;
                cont = true;
            }
        }
    }
};

Language.prototype.first = function(symbol) {
    if (symbol === '') return [];
    if (symbol instanceof Array) {
        var firsts = [];
        for (var i = 0, t; t = symbol[i]; ++i) {
            if (!this.nonterminals[t]) {
                if (firsts.indexOf(t) === -1) firsts.push(t);
            } else {
                union(firsts, this.nonterminals[t].first);
            }
            if (!this.nullable(t)) break;
        }
        return firsts;
    }
    return !this.nonterminals[symbol] ? [symbol] : this.nonterminals[symbol].first;
};

Language.prototype.followSets = function() {
    var productions = this.productions;
    var nonterminals = this.nonterminals;
    var self = this;
    var cont = true;

    while (cont) {
        cont = false;

        productions.forEach(function(production) {
            var q = !!self.go_;
            var ctx = q;

            for (var i = 0, t; t = production.handle[i]; ++i) {
                if (!nonterminals[t]) continue;

                if (ctx) {
                    q = self.go_(production.symbol, production.handle.slice(0, i));
                }
                var bool = !ctx || q === parseInt(self.lookahead.nterms_[t], 10);

                var set;
                if (i === production.handle.length - 1 && bool) {
                    set = nonterminals[production.symbol].follows;
                } else {
                    var part = production.handle.slice(i + 1);
                    set = self.first(part);
                    if (self.nullable(part) && bool) {
                        set.push.apply(set, nonterminals[production.symbol].follows);
                    }
                }

                var oldcount = nonterminals[t].follows.length;
                union(nonterminals[t].follows, set);
                if (oldcount !== nonterminals[t].follows.length) {
                    cont = true;
                }
            }
        });
    }
};

// LR operations
Language.prototype.closureOperation = function(itemSet) {
    var closureSet = new ItemSet();
    var self = this;
    var set = itemSet;
    var itemQueue, syms = {};

    do {
        itemQueue = [];
        closureSet.concat(set);

        (set._items || set).forEach(function(item) {
            var symbol = item.markedSymbol;

            if (symbol && self.nonterminals[symbol]) {
                if (!syms[symbol]) {
                    self.nonterminals[symbol].productions.forEach(function(production) {
                        var newItem = new Item(production, 0);
                        if (!closureSet.contains(newItem)) {
                            itemQueue.push(newItem);
                        }
                    });
                    syms[symbol] = true;
                }
            } else if (!symbol) {
                closureSet.reductions.push(item);
                closureSet.inadequate = closureSet.reductions.length > 1 || closureSet.shifts;
            } else {
                closureSet.shifts = true;
                closureSet.inadequate = closureSet.reductions.length > 0;
            }
        });

        set = itemQueue;
    } while (itemQueue.length > 0);

    return closureSet;
};

Language.prototype.gotoOperation = function(itemSet, symbol) {
    var gotoSet = new ItemSet();
    var self = this;

    itemSet._items.forEach(function(item, n) {
        if (item.markedSymbol === symbol) {
            gotoSet.push(new Item(item.production, item.dotPosition + 1, item.follows, n));
        }
    });

    return gotoSet.length === 0 ? gotoSet : this.closureOperation(gotoSet);
};

Language.prototype.canonicalCollection = function() {
    var item1 = new Item(this.productions[0], 0, [this.EOF]);
    var firstState = this.closureOperation(new ItemSet(item1));
    var states = [firstState];
    var marked = 0;
    var self = this;

    states.has = {};
    states.has[firstState] = 0;

    while (marked !== states.length) {
        var itemSet = states[marked];
        marked++;

        itemSet._items.forEach(function(item) {
            if (item.markedSymbol && item.markedSymbol !== self.EOF) {
                self.canonicalCollectionInsert(item.markedSymbol, itemSet, states, marked - 1);
            }
        });
    }

    return states;
};

Language.prototype.canonicalCollectionInsert = function(symbol, itemSet, states, stateNum) {
    var g = this.gotoOperation(itemSet, symbol);
    if (!g.predecessors) g.predecessors = {};

    if (g.length > 0) {
        var gv = g.valueOf();
        var i = states.has[gv];
        if (i === -1 || typeof i === 'undefined') {
            states.has[gv] = states.length;
            itemSet.edges[symbol] = states.length;
            states.push(g);
            g.predecessors[symbol] = [stateNum];
        } else {
            itemSet.edges[symbol] = i;
            states[i].predecessors[symbol].push(stateNum);
        }
    }
};

// Parse table generation
Language.prototype.parseTable = function(itemSets) {
    var states = [];
    var nonterminals = this.nonterminals;
    var operators = this.operators;
    var conflictedStates = {};
    var self = this;
    var s = 1, r = 2, a = 3;
    var NONASSOC = 0;

    itemSets.forEach(function(itemSet, k) {
        var state = states[k] = {};

        // Set shift and goto actions
        for (var stackSymbol in itemSet.edges) {
            itemSet._items.forEach(function(item) {
                if (item.markedSymbol === stackSymbol) {
                    var gotoState = itemSet.edges[stackSymbol];
                    if (nonterminals[stackSymbol]) {
                        state[self.symbols_[stackSymbol]] = gotoState;
                    } else {
                        state[self.symbols_[stackSymbol]] = [s, gotoState];
                    }
                }
            });
        }

        // Set accept action
        itemSet._items.forEach(function(item) {
            if (item.markedSymbol === self.EOF) {
                state[self.symbols_[self.EOF]] = [a];
            }
        });

        var allterms = self.lookAheads ? false : self.terminals;

        // Set reductions
        itemSet.reductions.forEach(function(item) {
            var terminals = allterms || self.lookAheads(itemSet, item);

            terminals.forEach(function(stackSymbol) {
                var action = state[self.symbols_[stackSymbol]];
                var op = operators[stackSymbol];

                if (action || (action && action.length)) {
                    var sol = self.resolveConflict(item.production, op, [r, item.production.id],
                        action[0] instanceof Array ? action[0] : action);
                    self.resolutions.push([k, stackSymbol, sol]);

                    if (sol.bydefault) {
                        self.conflicts++;
                        conflictedStates[k] = true;
                        if (self.options.noDefaultResolve) {
                            if (!(action[0] instanceof Array)) {
                                action = [action];
                            }
                            action.push(sol.r);
                        }
                    } else {
                        action = sol.action;
                    }
                } else {
                    action = [r, item.production.id];
                }

                if (action && action.length) {
                    state[self.symbols_[stackSymbol]] = action;
                } else if (action === NONASSOC) {
                    state[self.symbols_[stackSymbol]] = undefined;
                }
            });
        });
    });

    return states;
};

Language.prototype.resolveConflict = function(production, op, reduce, shift) {
    var sln = {
        production: production,
        operator: op,
        r: reduce,
        s: shift
    };
    var s = 1, r = 2, NONASSOC = 0;

    if (shift[0] === r) {
        sln.action = shift[1] < reduce[1] ? shift : reduce;
        if (shift[1] !== reduce[1]) sln.bydefault = true;
        return sln;
    }

    if (production.precedence === 0 || !op) {
        sln.bydefault = true;
        sln.action = shift;
    } else if (production.precedence < op.precedence) {
        sln.action = shift;
    } else if (production.precedence === op.precedence) {
        if (op.assoc === "right") {
            sln.action = shift;
        } else if (op.assoc === "left") {
            sln.action = reduce;
        } else if (op.assoc === "nonassoc") {
            sln.action = NONASSOC;
        }
    } else {
        sln.action = reduce;
    }

    return sln;
};

Language.prototype.findDefaults = function(states) {
    var defaults = {};
    states.forEach(function(state, k) {
        var i = 0, act;
        for (act in state) {
            if ({}.hasOwnProperty.call(state, act)) i++;
        }
        if (i === 1 && state[act][0] === 2) {
            defaults[k] = state[act];
        }
    });
    return defaults;
};

// LALR-specific methods
Language.prototype.lookAheads = function(state, item) {
    return (!!this.onDemandLookahead && !state.inadequate) ? this.terminals : item.follows;
};

Language.prototype.go = function(p, w) {
    var q = parseInt(p, 10);
    for (var i = 0; i < w.length; i++) {
        q = this.states[q].edges[w[i]] || q;
    }
    return q;
};

Language.prototype.goPath = function(p, w) {
    var q = parseInt(p, 10);
    var path = [];
    for (var i = 0; i < w.length; i++) {
        var t = w[i] ? q + ":" + w[i] : '';
        if (t) this.lookahead.nterms_[t] = q;
        path.push(t);
        q = this.states[q].edges[w[i]] || q;
        this.terms_[t] = w[i];
    }
    return { path: path, endState: q };
};

Language.prototype.buildNewGrammar = function() {
    var self = this;
    var newg = this.lookahead;

    this.states.forEach(function(state, i) {
        state._items.forEach(function(item) {
            if (item.dotPosition === 0) {
                var symbol = i + ":" + item.production.symbol;
                self.terms_[symbol] = item.production.symbol;
                newg.nterms_[symbol] = i;
                if (!newg.nonterminals[symbol]) {
                    newg.nonterminals[symbol] = new Nonterminal(symbol);
                }
                var pathInfo = self.goPath(i, item.production.handle);
                var p = new Production(symbol, pathInfo.path, newg.productions.length);
                newg.productions.push(p);
                newg.nonterminals[symbol].productions.push(p);

                var handle = item.production.handle.join(' ');
                var goes = self.states[pathInfo.endState].goes;
                if (!goes[handle]) {
                    goes[handle] = [];
                }
                goes[handle].push(symbol);
            }
        });
        if (state.inadequate) {
            self.inadequateStates.push(i);
        }
    });
};

Language.prototype.unionLookaheads = function() {
    var self = this;
    var newg = this.lookahead;
    var states = !!this.onDemandLookahead ? this.inadequateStates : this.states;

    states.forEach(function(i) {
        var state = typeof i === 'number' ? self.states[i] : i;
        if (state.reductions.length) {
            state.reductions.forEach(function(item) {
                var follows = {};
                for (var k = 0; k < item.follows.length; k++) {
                    follows[item.follows[k]] = true;
                }
                state.goes[item.production.handle.join(' ')].forEach(function(symbol) {
                    newg.nonterminals[symbol].follows.forEach(function(symbol) {
                        var terminal = self.terms_[symbol];
                        if (!follows[terminal]) {
                            follows[terminal] = true;
                            item.follows.push(terminal);
                        }
                    });
                });
            });
        }
    });
};

// Code generation
Language.prototype.generate = function(opt) {
    opt = Object.assign({}, this.options, opt);
    var moduleName = opt.moduleName || "parser";
    if (!moduleName.match(/^[A-Za-z_$][A-Za-z0-9_$]*$/)) {
        moduleName = "parser";
    }
    return this.generateCommonJSModule(opt);
};

Language.prototype.generateCommonJSModule = function(opt) {
    opt = Object.assign({}, this.options, opt);
    var moduleName = opt.moduleName || "parser";
    var out = this.generateModule(opt) +
        "\n\n\nif (typeof require !== 'undefined' && typeof exports !== 'undefined') {" +
        "\nexports.parser = " + moduleName + ";" +
        "\nexports.Parser = " + moduleName + ".Parser;" +
        "\nexports.parse = function () { return " + moduleName + ".parse.apply(" + moduleName + ", arguments); };" +
        "\nexports.main = function() {};" +
        "\nif (typeof module !== 'undefined' && require.main === module) {\n" +
        "  exports.main(process.argv.slice(1));\n}" +
        "\n}";
    return out;
};

Language.prototype.generateModule = function(opt) {
    opt = Object.assign({}, this.options, opt);
    var moduleName = opt.moduleName || "parser";
    var out = "/* parser generated by jison " + version + " */\n";
    out += (moduleName.match(/\./) ? moduleName : "var " + moduleName) +
        " = " + this.generateModuleExpr();
    return out;
};

Language.prototype.generateModuleExpr = function() {
    var module = this.generateModule_();
    var out = "(function(){\n";
    out += module.commonCode;
    out += "\nvar parser = " + module.moduleCode;
    out += "\n" + this.moduleInclude;
    out += "\nfunction Parser () {\n  this.yy = {};\n}\n" +
        "Parser.prototype = parser;" +
        "parser.Parser = Parser;" +
        "\nreturn new Parser;\n})();";
    return out;
};

Language.prototype.generateModule_ = function() {
    var tableCode = this.generateTableCode(this.table);
    var moduleCode = "{";
    moduleCode += [
        "trace: function trace() {}",
        "yy: {}",
        "symbols_: " + JSON.stringify(this.symbols_),
        "terminals_: " + JSON.stringify(this.terminals_).replace(/"([0-9]+)":/g, "$1:"),
        "productions_: " + JSON.stringify(this.productions_),
        "performAction: " + String(this.performAction),
        "table: " + tableCode.moduleCode,
        "defaultActions: " + JSON.stringify(this.defaultActions).replace(/"([0-9]+)":/g, "$1:"),
        "parseError: " + String(this.parseError),
        "parse: " + String(this.parse)
    ].join(",\n");
    moduleCode += "};";

    return {
        commonCode: tableCode.commonCode,
        moduleCode: moduleCode
    };
};

Language.prototype.generateTableCode = function(table) {
    // Generate compact JSON without spaces and with unquoted numeric keys
    var moduleCode = JSON.stringify(table, null, 0);

    // Remove quotes around numeric object keys
    moduleCode = moduleCode.replace(/"([0-9]+)"(?=:)/g, "$1");

    return {
        commonCode: '',
        moduleCode: moduleCode
    };
};

Language.prototype.parseError = function(str, hash) {
    if (hash.recoverable) {
        this.trace(str);
    } else {
        var error = new Error(str);
        error.hash = hash;
        throw error;
    }
};

Language.prototype.parse = function(input) {
    var self = this;
    var stack = [0];
    var vstack = [null];
    var lstack = [];
    var table = this.table;
    var yytext = '';
    var yylineno = 0;
    var yyleng = 0;
    var recovering = 0;
    var TERROR = 2;
    var EOF = 1;

    var lexer = Object.create(this.lexer);
    var sharedState = { yy: {} };

    for (var k in this.yy) {
        if (Object.prototype.hasOwnProperty.call(this.yy, k)) {
            sharedState.yy[k] = this.yy[k];
        }
    }

    lexer.setInput(input, sharedState.yy);
    sharedState.yy.lexer = lexer;
    sharedState.yy.parser = this;

    if (typeof lexer.yylloc === 'undefined') {
        lexer.yylloc = {};
    }
    var yyloc = lexer.yylloc;
    lstack.push(yyloc);

    var ranges = lexer.options && lexer.options.ranges;

    if (typeof sharedState.yy.parseError === 'function') {
        this.parseError = sharedState.yy.parseError;
    } else {
        this.parseError = Object.getPrototypeOf(this).parseError;
    }

    var lex = function() {
        var token = lexer.lex() || EOF;
        if (typeof token !== 'number') {
            token = self.symbols_[token] || token;
        }
        return token;
    };

    var symbol, preErrorSymbol, state, action, r, yyval = {};
    var p, len, newState, expected;

    while (true) {
        state = stack[stack.length - 1];

        if (this.defaultActions[state]) {
            action = this.defaultActions[state];
        } else {
            if (symbol === null || typeof symbol === 'undefined') {
                symbol = lex();
            }
            action = table[state] && table[state][symbol];
        }

        if (typeof action === 'undefined' || !action.length || !action[0]) {
            var errStr = '';

            if (!recovering) {
                expected = [];
                for (p in table[state]) {
                    if (this.terminals_[p] && p > TERROR) {
                        expected.push("'" + this.terminals_[p] + "'");
                    }
                }
                if (lexer.showPosition) {
                    errStr = 'Parse error on line ' + (yylineno + 1) + ":\n" +
                        lexer.showPosition() + "\nExpecting " + expected.join(', ') +
                        ", got '" + (this.terminals_[symbol] || symbol) + "'";
                } else {
                    errStr = 'Parse error on line ' + (yylineno + 1) + ": Unexpected " +
                        (symbol === EOF ? "end of input" :
                            ("'" + (this.terminals_[symbol] || symbol) + "'"));
                }
                this.parseError(errStr, {
                    text: lexer.match,
                    token: this.terminals_[symbol] || symbol,
                    line: lexer.yylineno,
                    loc: yyloc,
                    expected: expected
                });
            }

            throw new Error(errStr);
        }

        if (action[0] instanceof Array && action.length > 1) {
            throw new Error('Parse Error: multiple actions possible at state: ' + state + ', token: ' + symbol);
        }

        switch (action[0]) {
            case 1: // shift
                stack.push(symbol);
                vstack.push(lexer.yytext);
                lstack.push(lexer.yylloc);
                stack.push(action[1]);
                symbol = null;
                if (!preErrorSymbol) {
                    yyleng = lexer.yyleng;
                    yytext = lexer.yytext;
                    yylineno = lexer.yylineno;
                    yyloc = lexer.yylloc;
                    if (recovering > 0) {
                        recovering--;
                    }
                } else {
                    symbol = preErrorSymbol;
                    preErrorSymbol = null;
                }
                break;

            case 2: // reduce
                len = this.productions_[action[1]][1];
                yyval.$ = vstack[vstack.length - len];
                yyval._$ = {
                    first_line: lstack[lstack.length - (len || 1)].first_line,
                    last_line: lstack[lstack.length - 1].last_line,
                    first_column: lstack[lstack.length - (len || 1)].first_column,
                    last_column: lstack[lstack.length - 1].last_column
                };
                if (ranges) {
                    yyval._$.range = [lstack[lstack.length - (len || 1)].range[0],
                                      lstack[lstack.length - 1].range[1]];
                }
                r = this.performAction.apply(yyval, [yytext, yyleng, yylineno, sharedState.yy,
                                                     action[1], vstack, lstack]);

                if (typeof r !== 'undefined') {
                    return r;
                }

                if (len) {
                    stack = stack.slice(0, -1 * len * 2);
                    vstack = vstack.slice(0, -1 * len);
                    lstack = lstack.slice(0, -1 * len);
                }

                stack.push(this.productions_[action[1]][0]);
                vstack.push(yyval.$);
                lstack.push(yyval._$);
                newState = table[stack[stack.length - 2]][stack[stack.length - 1]];
                stack.push(newState);
                break;

            case 3: // accept
                return true;
        }
    }
};

Language.prototype.trace = function() {};

Language.prototype.createParser = function() {
    var p = eval(this.generateModuleExpr());
    p.productions = this.productions;

    var self = this;
    function bind(method) {
        return function() {
            self.lexer = p.lexer;
            return self[method].apply(self, arguments);
        };
    }

    p.lexer = this.lexer;
    p.generate = bind('generate');
    p.generateModule = bind('generateModule');
    p.generateCommonJSModule = bind('generateCommonJSModule');

    return p;
};

// Exports
Jison.Parser = function(grammar, options) {
    var gen = new Language(grammar, options);
    return gen.createParser();
};

exports.LALRGenerator = Language;

Jison.Generator = function(g, options) {
    var opt = Object.assign({}, g.options, options);
    return new Language(g, opt);
};

return function Parser(g, options) {
    var gen = Jison.Generator(g, options);
    return gen.createParser();
};
